{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dotenv import load_dotenv\n",
    "import os\n",
    "from GrobidArticleExtractor import GrobidArticleExtractor\n",
    "from langchain.prompts import PromptTemplate\n",
    "from entity_extraction import ChatOpenRouter, parse_pdf, get_model_output, process_and_save_output, create_chain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "load_dotenv()\n",
    "OPENROUTER_API_KEY = os.environ[\"OPENROUTER_API_KEY\"] \n",
    "GROBID_SERVER_URL = os.environ[\"GROBID_SERVER_URL\"]\n",
    "OPENAI_API_KEY = os.environ[\"OPENAI_API_KEY\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "grobid_client = GrobidArticleExtractor(GROBID_SERVER_URL)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "phillips_text = parse_pdf('data/Phillips_2023_11pg.pdf', grobid_client)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "llm_gpt_4o_mini = ChatOpenRouter('openai/gpt-4o-mini')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "llm_deepseek_v3 = ChatOpenRouter(model_name='deepseek/deepseek-chat-v3-0324:free')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "ner_prompt = PromptTemplate.from_template(\"Task: Perform Named Entity Recognition (NER) with a deep understanding of neuroscience-specific terminology. Use precise and domain-relevant entity labels.\\\n",
    "Return the output in a training-ready JSON object for spaCy NER training, provide the entity,start position, end position, and respective label.\\\n",
    "Here is an example output: \\\n",
    "```json 'text': 'Drug addiction is a chronic','entities': ['entity': 'Drug addiction','start': 0,'end': 13,'label': 'DISORDER']```\\\n",
    "The text is provided here: {neuroscience_text}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "ner_chain = create_chain(llm_deepseek_v3, ner_prompt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'heading': 'Introduction',\n",
       "  'content': 'Drug addiction is a chronic, complex neuropsychiatric disorder characterized by a loss of control of drug-taking behaviors. The current drug addiction and overdose epidemic in the United States has been worsened by the COVID-19 pandemic as people struggle with social isolation and economic distress (Cisneros and Cunningham, 2021). Much of the current addiction epidemic and surge in overdose deaths has been attributed to the use of opioids and, in particular, synthetic opioids such as fentanyl. However, drug overdose deaths associated with psychostimulants rose 50  between 2019 and 2020 (Hedegaard et al., 2021). Furthermore, drug overdose deaths associated with cocaine have increased 3-fold since 1999 (Hedegaard et al., 2021;Ciccarone, 2021). While pharmacological treatments exist for opioid use disorder, no such treatments are available for stimulant use disorder. Thus, continued research on the molecular adaptations that occur following exposure to cocaine, and the cell types that may be susceptible to these adaptations, is important for development of future therapeutics. The mesolimbic dopamine pathway is heavily implicated in molecular, cellular, and behavioral responses to drugs of abuse. This pathway consists of dopaminergic neurons in the ventral tegmental area (VTA) that project to the nucleus accumbens (NAc). Dopaminergic neurotransmission in the NAc is important for adaptive behavior, as dopamine (DA) release tracks reward-predictive cues as a function of learning (Day et al., 2007). Exposure to drugs of abuse, such as cocaine, increases extracellular concentrations of DA within the NAc. Cocaine extends DA transmission by blocking the DA transporter present on VTA dopaminergic neurons, and this mechanism is required for behavioral and physiological responses to cocaine (Giros et al., 1996). Drug-induced elevations in DA promote transcriptional and epigenetic reorganization that ultimately gives rise to the cellular and behavioral adaptations that drive addiction-linked behaviors. Within the NAc, DA engages the type-1 DA receptor (DRD1, encoded by the Drd1 gene) that is present on a subset of medium spiny neurons (MSNs), called D1-MSNs. DRD1 activation stimulates signal transduction cascades that results in CREB phosphorylation and activation of immediate early gene (IEG) expression programs. These IEG programs are comprised largely of activitydependent transcription factors such as Fos, FosB, and Nr4a1 (Savell et al., 2020;Nishi et al., 2011;Bertran-Gonzalez et al., 2008), each of which are critical mediators for cellular and behavioral adaptations to drugs of abuse (Zhang et al., 2006;Nestler et al., 2001;Carpenter et al., 2020). The NAc is a heterogeneous brain region consisting of many transcriptionally defined cell types. While a significant amount of research has focused on the role of DA receptor-expressing MSNs in adaptations to drugs of abuse, recent reports have also demonstrated significant roles for other cell types. For example, optogenetic inhibition of NAc somatostatin expressing interneurons (Sst-interneurons) attenuates cocaine-induced conditioned place preference (Ribeiro et al., 2018), suggesting that these cells are important for behavioral adaptations to drugs of abuse. Similarly, photoinhibition of cholinergic interneurons (Chat-interneurons) also blocks cocaine-induced conditioned place preference (Witten et al., 2010). Interestingly, recent research has also demonstrated an important role for astrocytes in mediating cocaineinduced synaptic and behavioral adaptations (Wang et al., 2021;Corkrum et al., 2020). These results demonstrate that diverse cell types within the NAc contribute to the molecular, cellular, and behavioral effects of cocaine. Despite established cell type heterogeneity in the NAc, many studies rely on transgenic animal models that allow for the investigation of these changes only in individual cell populations. Unlike many techniques focused on a defined class of cells, single nucleus RNA-sequencing (snRNA-seq) provides unprecedented resolution that allows for the investigation of thousands of genes in thousands of cells with a single assay. This resolution not only permits detection of cocaine-induced transcriptional changes in many cell types, but also allows for better understanding of the molecular heterogeneity within the NAc. Recently, snRNA-seq approaches have been used to generate cellular atlases of the mouse (Chen et al., 2021), non-human primate (He et al., 2021), and human (Tran et al., 2021) NAc. However, snRNA-seq has only been used to define early gene expression changes to acute cocaine, while the transcriptional effects of repeated cocaine in various NAc cell types is less well studied. Here, we used snRNA-seq to transcriptionally profile 39,254 nuclei from male and female Sprague-Dawley rats. Rats were further subdivided by treatment (cocaine or saline) and exposure paradigm (acute or repeated). This study confirms both the presence of many previously defined cell types within the rat NAc, and reveals that D1-MSNs exhibit the most robust transcriptional response to cocaine across all cell populations. Additionally, snRNA-seq of the rat NAc allowed for the investigation of cluster-specific expression of different neurotransmitter receptor systems and identified novel marker genes for two populations of transcriptionally distinct D1-MSNs.',\n",
       "  'subsections': []},\n",
       " {'heading': 'Identification of transcriptionally distinct cell types within the rat NAc',\n",
       "  'content': 'To begin understanding cell-type heterogeneity within the rat NAc, we first integrated 15,655 nuclei from a previously published dataset (Savell et al., 2020) consisting of male and female adult Sprague-Dawley rats that received a single (acute) injection of cocaine with 23,599 nuclei from a new dataset following seven daily (repeated) injections of cocaine (Fig. 1a, Supp. Fig. 1). Following integration and dimensionality reduction, we identified 11 transcriptionally distinct neuronal populations and 5 glial populations that were not specific to a single dataset and did not differ by sex (Fig. 1b,c,Supp. Figs. 2,3). Furthermore, calculation of Local Inverse Simpsons Index (LISI) demonstrated that all clusters were well mixed, indicating successful integration (Supp. Fig. 2). A significant portion of neurons within the NAc are GABAergic MSNs that express high levels of Gad1, Ppp1r1b,Foxp2,and Bcl11b (Fig. 1d,e). A large majority of MSNs can be further classified by the expression of Drd1, the gene encoding the type-1 dopamine receptor, or Drd2, the gene encoding the type-2 dopamine receptor (Fig. 1d). While we have previously reported the presence of two transcriptionally distinct Drd2-expressing MSN populations (Drd2-MSN-1 and Drd2-MSN-2) (Savell et al., 2020), this analysis also identified two transcriptionally distinct Drd1-expressing MSN populations as well (Drd1-MSN-1 and Drd1-MSN-2; Fig. 1b,d). In addition to Drd1-and Drd2-MSNs, this analysis also identified previously reported Drd3-and Grm8-expressing MSNs (Drd3-MSN and Grm8-MSN), which have also been identified within the striatum of non-human primates (He et al., 2021). While most of the GABAergic neurons in the NAc are MSNs, distinct populations of GABAergic interneurons are also known to play critical yet distinct roles in striatal function (Kawaguchi, 1993). While we have previously characterized parvalbumin and somatostatin interneurons within the NAc (Savell et al., 2020), this analysis also identified a cholinergic interneuron population marked by the expression of Chat (Chat-Interneuron). Previous research has demonstrated that Chat-Interneurons are critical regulators of drug-induced behavioral adaptations (Witten et al., 2010), and this dataset now provides comprehensive transcriptional profiling of these cells within the rat NAc. In addition to transcriptionally distinct neuronal populations, we also identified 5 glial clusters that are almost entirely devoid of Gad1 and Syt1 mRNA, and expressed canonical markers of their respective populations (Fig. 1d,e). For example, Astrocytes exhibit high expression of both Aqp4 and Gja1, while Oligodendrocytes exhibit high expression of Mbp, the gene encoding myelin basic protein (Fig. 1d,e).',\n",
       "  'subsections': []},\n",
       " {'heading': 'Computational investigation of transcriptionally distinct populations of Drd1-expressing MSNs',\n",
       "  'content': 'The integration of over 39,000 nuclei from rats that underwent different cocaine exposure paradigms identified two populations of Drd1-expressing MSNs (Drd1-MSN-1 and Drd1-MSN-2; Fig. 2a). To begin understanding the distinct gene expression signatures of these Drd1-MSN populations, we conducted a differential expression analysis using pseudobulked gene expression matrices. This analysis identified 506 genes enriched in the Drd1-MSN-1 cluster and 665 genes enriched in the Drd1-MSN-2 cluster. Interestingly, Ebf1 is the most highly enriched differentially expressed gene in the Drd1-MSN-1 population (Fig. 2b). Ebf1 is a known marker of D1-MSNs and is implicated in MSN early development (Lobo et al., 2008). Calb1, the gene encoding the calcium binding protein Calbindin 1, is another gene that is enriched in the Drd1-MSN-1 population (Fig. 2b). Calb1 is a spatial marker within the NAc and is highly enriched in the NAc core (Jongen-Rêlo et al., 1994). Included in the top 10 differentially expressed genes in the Drd1-MSN-2 population was Htr4, the gene that codes for the serotonin receptor 4 (Fig. 2b). HTR4 autoradiography suggests that this receptor is located in the ventral portion of the NAc shell (Jean et al., 2012). The differential enrichment of Htr4 and Calb1 suggests that the two populations of Drd1expressing MSNs identified by snRNA-seq are both transcriptionally and spatially distinct. Next, we set out to understand the specific gene networks and pathways that may be enriched in these two D1-MSN populations. Molecular function pathway analysis revealed an enrichment of genes involved in calcium transport and signaling in the Drd1-MSN-1 population, but enrichment of genes involved in potassium transport in the Drd1-MSN-2 population (Fig. 2c). For example, Trpc3, a gene encoding a calcium permeable channel, and Cacng3, a gene encoding an L-type calcium channel, are both highly enriched in the Drd1-MSN-1 population (Fig. 2d). Conversely, several genes encoding potassium channels, such as Kcnc1, Kcnc2, and Kcng3, are enriched in the Drd1-MSN-2 population, further highlighting the enrichment of genes involved in potassium transport in this population. While these genes are differentially enriched, both calcium-and potassium-related genes are expressed to some extent in both clusters (Fig. 2d). Functional investigation of transcriptionally distinct populations of neurons in the mammalian brain requires the use of cell type-specific promoters. To facilitate the future investigation of the two populations of transcriptionally distinct Drd1-expressing MSNs, we next identified potential marker genes that are both differentially and highly expressed within a population. This analysis identified Ebf1 as a potential marker gene for the Drd1-MSN-1 population (Fig. 2e) and Htr4 as a potential marker gene for the Drd1-MSN-2 population (Fig. 2f). An additional characteristic of marker genes is that they are expressed in a large proportion of cells within the cluster. Plotting the expression of these genes on a UMAP shows differential enrichment of Ebf1 and Htr4 between the Drd1-expressing MSN populations (Fig. 2g). However, Ebf1 is also expressed in endothelial cells, while Htr4 is also expressed in Drd2and Drd3-MSNs. Thus, these genes may act as marker genes in combination with Drd1 but cannot act as marker genes alone.',\n",
       "  'subsections': []},\n",
       " {'heading': 'In-situ validation of transcriptionally distinct Drd1-expressing MSNs',\n",
       "  'content': 'To validate snRNA-seq findings demonstrating that Htr4 and Ebf1 are in largely distinct populations of Drd1-expressing MSNs in the NAc, we used RNAscope (Wang et al., 2012), a previously established fluorescence in situ hybridization (FISH) protocol. To enable visualization of distinct puncta representing cells expressing a combination of Drd1, Htr4, and Ebf1, we performed multiplexed analysis with probe sets specific to each transcript (Fig. 3a-c). Consistent with our computational analysis, we identify unique subpopulations of Drd1 cells that are either Ebf1 or Htr4, with significantly fewer cells that are positive for both Ebf1 and Htr4 (Fig. 3b-c). Furthermore, our in situ experiments revealed unique spatial distributions for these subpopulations with a significantly higher proportion of Drd1Htr4 cells in the NAc shell and Drd1Ebf1 cells in the NAc core (Fig. 3d-e). There was no significant spatial distribution difference in the proportion of Drd1 Ebf1Htr4 cells. Overall, these results are consistent with snRNA-seq analysis, and confirm largely non-overlapping expression of Ebf1 and Htr4 within Drd1 cells (Fig. 3f).',\n",
       "  'subsections': []},\n",
       " {'heading': 'Cocaine experience differentially alters the transcriptome of specific cell types within the NAc',\n",
       "  'content': 'We previously demonstrated that a single acute cocaine exposure differentially alters the transcriptome of molecularly defined cell types within the NAc (Savell et al., 2020). To begin understanding how repeated cocaine exposure alters the transcriptome of distinct NAc populations, we injected male and female Sprague-Dawley rats with 20 mgkg cocaine for 7 consecutive days (Fig. 1a, Supp. Fig. 1). Before directly comparing acute and repeated exposures, we first investigated cell type-specific transcriptional responses to cocaine exposure, regardless of type or duration. To do this, we collapsed both datasets and used a pseudobulk approach to identify cocaine-dependent, cell typespecific differentially expressed genes (DEGs; Supp. Table 2). This approach increased the number of replicates used for statistical comparisons and circumvented previously identified issues with deprecated differential expression methods previously used for snRNA-seq (Squair et al., 2021). Strikingly, the Drd1-MSN-1 population exhibited the most robust transcriptional response to cocaine, with 236 total DEGs following cocaine experience (184 upregulated, 52 downregulated; Fig. 4a,b). These upregulated DEGs include immediate early genes (IEGs) such as Nr4a3, Tiparp, and Fosl2 (Fig. 4b). Recent evidence has demonstrated that astrocytes within the NAc also play a key role in the neuronal and behavioral response to dopamine and drugs of abuse (Wang et al., 2021;Corkrum et al., 2020;Shelkar et al., 2022). Disruption of glutamatergic transport, release, and signaling is implicated in substance use disorders (Reissner and Kalivas, 2010;Kalivas, 2009), and astrocytes are heavily involved in the regulation of this process across the brain (Mahmoud et al., 2019). Interestingly, Gria1, an ionotropic AMPA receptor, is significantly decreased in Astrocytes following cocaine exposure (Fig. 4c). Unlike the Drd1-MSN-1 population, Astrocytes exhibited more genes with cocaine-dependent decreases in expression (Fig. 4c,g). While most DEGs in the Astrocyte population are downregulated with cocaine, Pde10a, the gene encoding the cAMPregulated phosphodiesterase 10A, is upregulated by cocaine in both the Astrocyte and Sst-Interneuron cluster (Fig. 4c,d). Previous research has demonstrated that Drd1-expressing MSNs exhibit robust transcriptional changes in response to cocaine (Savell et al., 2020;Bertran-Gonzalez et al., 2008;Zhang et al., 2006;Zhang et al., 2005). More specifically, snRNA-seq from the rat NAc following acute cocaine exposure demonstrated that D1-MSNs have the most robust transcriptional response to cocaine when compared to other transcriptionally distinct NAc cell types (Savell et al., 2020). In this dataset, only one of the Drd1-expressing MSN populations (Drd1-MSN-1) had a robust transcriptional response to cocaine. For example, Nr4a3, Fosb, Fosl2, and Tiparp are all significantly increased following cocaine exposure only in the Drd1-MSN-1 population (Fig. 4e). Next, we set out to investigate the relationship of cocaine dependent DEGs across cocaine exposure paradigms. To do this, we first conducted DEG analyses comparing cocaine and saline samples with exposure type ignored (i.e., all cocaine samples vs all saline samples), acute exposure only, and repeated exposure samples only. The log2 (Fold Change) values for all DEGs were then concatenated and subject to hierarchical clustering. Overall, this analysis identified that many of the genes altered by acute cocaine are also altered by repeated cocaine in the same direction (Fig. 4f-h). However, for the Drd1-MSN-1 population some genes altered by acute cocaine are not altered by repeated cocaine. Furthermore, while the overall pattern shows concordant gene expression patterns, some cocaine-regulated genes within the Astrocyte cluster demonstrate opposite changes with an acute or repeated exposure. To estimate the degree of concordance between gene expression programs altered by acute and repeated cocaine exposure in a thresholdfree manner, we next used the rank-rank hypergeometric overlap (RRHO) algorithm (Cahill et al., 2018). To complete this analysis, we ranked gene lists by the log2 (Fold Change), which was calculated by dividing the log-normalized expression values of cocaine cells by the log-normalized expression values of saline cells. This analysis identified that, on a cell type-specific basis, acute and repeated cocaine gene expression signatures are primarily concordant. For example, the acute and repeated cocaine changes for the Drd1-MSN-1, Astrocyte, and Sst-Interneuron populations are significantly concordant for both upregulated and downregulated genes (Fig. 4i-l). However, the strength of this overlap and concordance is substantially different for these populations.',\n",
       "  'subsections': []},\n",
       " {'heading': 'Conservation of MSN subtypes across mammalian species',\n",
       "  'content': 'Recently, transcriptional atlases of the non-human primate (NHP) striatum (He et al., 2021) and human NAc (Tran et al., 2021) have demonstrated similar gene expression profiles of specific marker genes that we identify within the rat NAc. To quantify this overlap and perform cross-species comparisons, we focused on conservation of MSNs given that both NHP and human datasets identified and validated the presence of several transcriptionally distinct MSN subtypes. Direct comparison of our rat snRNA-seq dataset to monkey or human datasets revealed potential conservation of many NHP and human cell types (Fig. 5a-c). For example, the rat Drd1-MSN-1 population is correlated with the human D1.MSN_A and NHP D1.Matrix populations, and the rat Drd2-MSN-1 population is correlated with the human D2.MSN_A and the NHP D2. Matrix population (Fig. 5a-d). Transcriptionally conserved cell types should exhibit similar expression and enrichment of marker genes. For example, Drd2-MSNs are classically characterized by the enrichment of Drd2 and Penk, while Drd1-MSNs are characterized by the enrichment of Drd1 and Ebf1. Across species, transcriptionally distinct populations of Drd2-MSNs are similarly enriched for Drd2, but differ in their enrichment of Penk (Fig. 5c, Supp. Fig. 4a-c). A similar pattern is observed in Drd1-MSNs across species, as Drd1 is highly enriched in Drd1-MSNs, but Ebf1 enrichment differs (Fig. 5c, Supp. Fig. 4a-c). While the differential expression of Htr4 or Ebf1 distinguishes Drd1-MSNs in the rat NAc, this pattern is not observed in higher order mammals as Htr4 is ubiquitously expressed across cell types (Supp. Fig. 4a-c). Furthermore, Htr4 is enriched in the MSN.D1_A human population (Fig. 5c) along with Ebf1. Thus, Htr4 and Ebf1 may not be suitable markers for distinguishing between transcriptionally distinct populations of Drd1-MSNs in higher order mammals.',\n",
       "  'subsections': []},\n",
       " {'heading': 'Discussion',\n",
       "  'content': 'Within the past few decades, a major focus of drug addiction research has been on the transcriptional and epigenetic changes occurring in the NAc following drug exposure. While these studies have been integral in our understanding of how drug experience shapes the brain, they have Fig. 5. Identification of conserved neuronal cell types across higher order mammals. a-b, Heatmap colored by the Pearsons correlation coefficient between rat, nonhuman primate, and human cell types. Correlation coefficients were calculated using t-statistic for rat MSN DEGs and their orthologs. c, Heatmaps colored by tstatistic for specific MSN marker genes in rat, non-human primate, and human. d, Estimated conservation of NAc neuronal cell types across rat, non-human primate, and human. primarily used techniques that ignore cell-type heterogeneity. Furthermore, although the NAc is fundamental in reward processing and the rat is a primary model organism used for reward-related research, few studies have comprehensively profiled the rat NAc at the transcriptional level. snRNA-seq provides the unique opportunity to study both the cell type-specific transcriptional responses to drugs of abuse, as well as the cell-type heterogeneity that exists within the NAc. In the present study, we report the integration of a previously published NAc snRNA-seq dataset collected following acute cocaine exposure (Savell et al., 2020) with a new dataset collected following repeated cocaine exposure. The resulting dataset contains 39,254 nuclei across 16 transcriptionally distinct populations (11 neuronal and 5 non-neuronal), which more than doubles the size of our previously published work. The addition of both biological and technical replicates, along with the addition of over 24,000 nuclei, allowed us to generate a comprehensive transcriptional atlas of the rat NAc and investigate cell type-specific transcriptional responses to cocaine. We were also interested in understanding the distribution of expression of genes encoding different receptor systems, which provides additional information regarding which cell types may respond to diverse classes of drugs. For example, there is significant renewed interest in the action of psychedelic drugs, many of which act on serotonin receptors. Htr2a, the gene encoding serotonin receptor type 2A and the expected target of many psychedelics, is expressed in a small proportion of Drd1-and Drd2-expressing MSNs, as well as Sst-interneurons (Supp. Fig. 5a). Thus, these populations of cells will most likely respond to classic psychedelics. There is also considerable interest in the distribution of expression of opioid receptors in the NAc as opioid-associated overdose deaths continue to increase (Hedegaard et al., 2021;Ciccarone, 2021). Previous research has demonstrated that 25  of cells containing the delta opioid receptor (Oprd1) in the NAc also contained Met 5 -Enkaphalin (Svingos et al., 1998), a peptide encoded by the gene Penk. Penk is a marker gene for Drd2-expressing MSNs, suggesting that the delta opioid receptor is expressed in this population of neurons. Our data confirms this observation, as Oprd1 is highly expressed in the Drd2-MSN-1 population (Supp. Fig. 5a-c). A major focus of this study was to characterize cell type-specific gene programs induced by cocaine experience. Prior work has revealed distinct differences in transcriptomic changes between volitional cocaine paradigms, such has self-administration, and experimenteradministered injections. For example, after a prolonged withdrawal period, animals that underwent cocaine self-administration exhibit increased surface expression of calcium permeable AMPA receptors, while animals that underwent experimenter-administered injections do not (McCutcheon et al., 2011). Additionally, a recent study systematically tested transcriptomic changes occurring in the VTA across several cocaine paradigms and identified several IEGs that are upregulated by experimenter-administered cocaine injections (Campbell et al., 2021). Like this study, we also identified several IEGs (Fosl2, Tiparp, Fosb), that are upregulated by experimenter-administered cocaine injections. Notably, we did not observe significant locomotor sensitization to repeated doses of cocaine in the present study (Supp. Fig. 1b). While sensitization is commonly (but not always (Gulley et al., 2003;Mayfield et al., 1992;Cass et al., 1993)) observed with repeated cocaine injections, it is possible that distinct gene expression patterns may be observed in animals that exhibit sensitized responses to cocaine as compared to those that do not. This will be an interesting consideration for future studies that are able to obtain single-nuclei transcriptional profiles while preserving individual animal variability in the data structure. Ultimately, the results presented in the current study shed light on cocaine-induced, cell-type specific transcriptomic changes occurring in the NAc, expanding our understanding of how drug experience shapes the brain. Further research investigating paradigmspecific transcriptional changes may identify molecular correlates of more complex behaviors, such as cocaine seeking or craving. The largest transcriptional response to cocaine experience occurred in the Drd1-MSN-1 population (Fig. 4a,b). These changes included IEGs such as Fosl2. IEGs and other cocaine-regulated genes in the Drd1-MSN-1 population are induced by both acute and repeated cocaine, but to a lesser extent with repeated cocaine. This data is in agreement with previous research that demonstrated induction of FOSB protein by repeated cocaine was weaker as compared to acute cocaine (Zhang et al., 2005). The second population of Drd1-expressing MSNs within the NAc, Drd1-MSN-2, did not have a robust transcriptional response to cocaine experience. A multitude of previous research has demonstrated that Drd1-expressing cells within the NAc exhibit transcriptional and epigenetic responses to drugs of abuse (Savell et al., 2020;Bertran-Gonzalez et al., 2008;Zhang et al., 2006;Zhang et al., 2005;Zipperly et al., 2021). Thus, the identification of a population of Drd1-expressing MSNs in the NAc with a significantly attenuated transcriptional response to cocaine is intriguing because it highlights a previously underappreciated level of molecular heterogeneity within this population of cells. Notably, further analysis of population-specific DEGs also revealed an enrichment of calcium transport-related terms for Drd1-MSN-1 and potassium transport-related terms for Drd1-MSN-2. The differential enrichment for calcium-related terms and genes may explain the transcriptional response to cocaine in the Drd1-MSN-1 population, as many IEGs rely on calcium-dependent signaling cascades for induction. Our dataset also provided the opportunity to identify marker genes for these two populations. We identified Ebf1 as a marker gene for the Drd1-MSN-1 population and Htr4 as a marker gene for the Drd1-MSN-2 population. In situ experiments validated the presence of two distinct populations of Drd1-expressing MSNs based on the presence of Htr4 and Ebf1. These in situ experiments also revealed a spatial separation of these MSNs, as Drd1Ebf1 cells were primarily found in the NAc core and Drd1Htr4 cells were primarily found in the NAc shell. Transcriptional similarity based on spatial location may also explain why the Drd1-MSN-1 population is more closely related to the Drd2-MSN-1 population than the Drd1-MSN-2 population (Supp. Fig. 4a). In fact, analysis of the relationship between cell types within the NHP identifies that Drd1-and Drd2-expressing MSNs from the same striatal subregions are more closely related than Drd1-expressing MSNs found in distinct striatal compartments (Supp. Fig. 4b). Similarly, the MSN.D1_A and MSN.D2_A human populations are more closely related to each other than other Drd1-or Drd2-expressing MSNs, although no spatial information is available in that dataset (Supp. Fig. 4c). Together, these results suggest that at least some component of transcriptional diversity in the NAc reflects the spatial organization of this structure. Although sex differences in drug-induced behavioral adaptations have long been appreciated (Hu et al., 2004;Bobzean et al., 2014;Becker and Chartoff, 2019;Evans, 2007), more recent research has also demonstrated sex-specific drug-induced molecular adaptations (López et al., 2021). Our dataset includes nuclei from both male and female rats, and thus we examined sex-specific transcriptional responses to cocaine across cell populations within the NAc, again using the collapsed acute and repeated dataset to enable pseudobulk treatment comparisons between male and females. In general, the direction of cocaine-induced mRNA changes was largely concordant between males and females across most cell populations (Supp. Fig. 6a-d), however significance criteria were often only met for one sex (Supp. Table 3). Overall, the Drd1-MSN-1 population was the most transcriptionally responsive to cocaine for both sexes in terms of total number of genes altered. Consistent with the sex-combined identification of cocaine DEGs, immediate early gene Nr4a3, transcription elongation factor Ell2, and cytoskeletal regulator Elmo2 were all significantly elevated in the Drd1-MSN-1 population in both sexes (Supp. Fig. 6e). Interestingly, sexspecific cocaine DEG comparisons revealed that while some genes, like Nr4a3 and Ell2, display relatively consistent patterns of expression between males and females across populations (Supp. Fig. 6e-f), other genes like Hs6st3 exhibit more dynamic patterns of expression (Supp. Fig. 6g). In conclusion, we have generated a comprehensive transcriptional atlas of the rat NAc that also provides information regarding cocainedependent gene expression changes. The use of snRNA-seq provided the ability to understand cell types that potentially mediate cellular and behavioral responses to drugs of abuse. Analysis of this dataset resulted in the identification of transcriptionally distinct populations of Drd1expressing MSNs that differ by their spatial location and transcriptional response to cocaine. This dataset will be a resource for future research into transcriptionally distinct populations of NAc cells and their responses to drugs of abuse.',\n",
       "  'subsections': []},\n",
       " {'heading': 'Behavioral testing',\n",
       "  'content': 'All experiments were performed in accordance with the University of Alabama at Birmingham Institutional Animal Care and Use Committee. Male or female adult rats were intraperitoneally injected with Cocaine hydrocholoride at a dose of 20 mgkg or saline. Locomotor testing was performed as previously described (Savell et al., 2020). Animals undergoing acute cocaine exposure received a single dose, while those undergoing repeated cocaine exposure received one injection per day for 7 days.',\n",
       "  'subsections': []},\n",
       " {'heading': 'Single-nuclei dissociation',\n",
       "  'content': 'Nuclei dissociation for snRNA-seq was performed as previously described (Savell et al., 2020). Nuclei from 4 animals of the same sex, treatment group, and exposure length were pooled for tissue preparation and loaded into a single GEM (gel beads in emulsion) well.',\n",
       "  'subsections': []},\n",
       " {'heading': 'snRNA-seq analysis methods',\n",
       "  'content': 'Libraries were constructed according to manufacturers instructions using the Chromium Single Cell 3  Library Construction Kit (10 Genomics, catalog no. 1000092), which uses version 3 chemistry for gene expression. Nuclei (15,655) from adult rat NAc after acute cocaine and nuclei (23,599) from adult rat NAc after repeated cocaine were sequenced on the Illumina NextSeq 500 at the UAB Heflin Genomics core to a depth of 39,000 reads per nuclei and 22,000 reads per nuclei, respectively. All raw fastq files were aligned to the Ensembl mRatBn7.2 (Rn7) genome using CellRanger (v6.1.2) with the associated Ensembl gene transfer format (gtf) file (version 105). Cell Ranger filtered outputs were analyzed with Seurat v4.0.4 (Hao et al., 2021) using R v4.0.2. Nuclei containing 200 features and 5  of reads mapping to the mitochondrial genome were removed from further analysis. Molecular count data from each GEM (gel beads in emulsion) well were then log-normalized with a scaling factor of 10,000. Following normalization, and before dimensionality reduction, all GEM wells were then integrated with FindIntegrationAnchors() and IntegrateData using 17 principal components (PCs) and a resolution value of 0.1 for both the acute and repeated datasets. Local Inverse Simpsons Index was calculated to ensure successful integration of all 8 GEM wells using lisi v1.0 with default parameters. Recent research has demonstrated that ambient RNA can be captured with single cells or nuclei with dropletbased single-cellsingle-nucleus technologies. To avoid interpreting any signal associated with ambient RNA, we used SoupX v1.5.2 (Young and Behjati, 2020) to remove any signal associated with ambient RNA. Droplet based single-cellsingle-nucleus technologies are also susceptible to the capture of doublets, or two cellsnuclei being captured within a single droplet. Homotypic doublets, or doublets containing two cells of the same cell-type, are difficult to identify and remove. However, heterotypic doublets, or doublets containing two cells of different cell types, can be identified as cells that have a combination of cell typespecific transcriptional signatures. Heterotypic doublets were not identified in our acute dataset. However, heterotypic doublets were identified and removed from the repeated dataset using DoubletFinder v2.0.3 (McGinnis et al., 2019). For differential expression testing, gene counts were first summed across GEM wells within each cell type to create a pseudobulked gene expression matrix. Any genes with counts 5 were removed from further analysis. Differential expression testing was then completed using pseudobulked gene expression matrices with DESeq2 v1.28.1 (Love et al., 2014) using the Likelihood Ratio Test (LRT). The data sets utilized in this paper are from two separate experiments and therefore represent two separate batches. Batch effects can substantially alter transcriptional profiles and to control for this data set was included as a covariate in all differential expression testing. In addition to data set, treatment was used as a covariate for differential expression testing across genetic sex and cell types. Including these variables as covariates allows us to ensure that all DEGs are due to treatment, cell type, or sex and not another confounding variable. Gini coefficients were calculated using the gini() function that is available in the reldist v1.6-6 package (Gini, 1921;Handcock and Morris, 1999). Rank-rank hypergeometric overlap (RRHO) was calculated using the log2 (Fold Change) of all genes regardless of adjusted p-value. RRHO heatmaps were generated using RRHO2 v1.0 (Cahill et al., 2018;Plaisier et al., 2010). Molecular function gene ontology terms were generated with DEGs expressed in 10  of the cluster using g:GOSt made available by gProfiler (Raudvere et al., 2019). For this tool, the statistical domain scope set to Only annotated genes and significance threshold set to Benjamini-Hochberg FDR. All R code for this analysis is available at https:github.comJeremy-Day-LabPhillips_2023_NAc.',\n",
       "  'subsections': []},\n",
       " {'heading': 'Calculation of t-statistic for conservation analysis',\n",
       "  'content': 'To correlate gene expression signatures across mammalian species, DEGs for rat cell types were first calculated using a pseudobulk approach. DEGs were then concatenated and t-statistics for every gene in every MSN subtype were calculated. The t-statistic was calculated with the following formula:  Following calculation of t-statistics, the Pearson correlation coefficient was calculated using the cor() function in R. All R code for this analysis is available at https:github.comJeremy-Day-LabPhillips_2023_NAc.',\n",
       "  'subsections': []},\n",
       " {'heading': 'Tissue preparation and RNAscope',\n",
       "  'content': 'Upon sacrifice, brains from 6 drug-naïve Sprague-Dawley rats (3F 3M) were extracted and immediately placed in isopentane at -40  C for 30 s. Tissue was stored at -80  C until day of sectioning. On day of sectioning, tissue was allowed to equilibrate at -20  C for 40 min and 10 μm coronal sections containing the NAc (1.92 mm to 1.44 mm from bregma) were collected using Leica cryostat and mounted on room temperature slides. Sections were air dried on dry ice for 60 min, then stored at -80  C. Sections were prepared for RNAscope Multiplex Fluorescent v2 Assay according to manufacturers protocol (Advanced Cell Diagnostics, 323110) (Wang et al., 2012). Probes included Htr4-C2 (544081-C2), and Drd1a-C3 (317031-C3) diluted 1:50 in Ebf1-C1 (891561-C1) and DAPI (320858). Associated Opal dyes were 690 (FP1497001KT) for Ebf1-C1, 570 (FP1488001KT) for Htr4-C2, and 520 (FP1487001KT) for Drd1a-C3 diluted 1:500 in TSA buffer. Sections were imaged using Keyence BZ-X800 microscope with OP-87767 filter cubes for GFP, Texas Red PE, DAPI, and Cy5 at 20 (PlanFluor NA 0.75) and',\n",
       "  'subsections': []}]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phillips_text.get('sections')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Running NER on sections: 100%|██████████| 12/12 [12:36<00:00, 63.00s/it]\n"
     ]
    }
   ],
   "source": [
    "all_output = []\n",
    "sections = phillips_text.get('sections', [])\n",
    "\n",
    "for section in tqdm(sections, desc=\"Running NER on sections\"):\n",
    "    chunk = section.get('content')\n",
    "    output = ner_chain.invoke({\"neuroscience_text\": chunk})\n",
    "    all_output.append(output)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[AIMessage(content='Here is the spaCy NER training-ready JSON output with neuroscience-specific entity labels:\\n\\n```json\\n{\\n  \"text\": \"Drug addiction is a chronic, complex neuropsychiatric disorder characterized by a loss of control of drug-taking behaviors. The current drug addiction and overdose epidemic in the United States has been worsened by the COVID-19 pandemic as people struggle with social isolation and economic distress (Cisneros and Cunningham, 2021). Much of the current addiction epidemic and surge in overdose deaths has been attributed to the use of opioids and, in particular, synthetic opioids such as fentanyl. However, drug overdose deaths associated with psychostimulants rose 50% between 2019 and 2020 (Hedegaard et al., 2021). Furthermore, drug overdose deaths associated with cocaine have increased 3-fold since 1999 (Hedegaard et al., 2021; Ciccarone, 2021). While pharmacological treatments exist for opioid use disorder, no such treatments are available for stimulant use disorder. Thus, continued research on the molecular adaptations that occur following exposure to cocaine, and the cell types that may be susceptible to these adaptations, is important for development of future therapeutics. The mesolimbic dopamine pathway is heavily implicated in molecular, cellular, and behavioral responses to drugs of abuse. This pathway consists of dopaminergic neurons in the ventral tegmental area (VTA) that project to the nucleus accumbens (NAc). Dopaminergic neurotransmission in the NAc is important for adaptive behavior, as dopamine (DA) release tracks reward-predictive cues as a function of learning (Day et al., 2007). Exposure to drugs of abuse, such as cocaine, increases extracellular concentrations of DA within the NAc. Cocaine extends DA transmission by blocking the DA transporter present on VTA dopaminergic neurons, and this mechanism is required for behavioral and physiological responses to cocaine (Giros et al., 1996). Drug-induced elevations in DA promote transcriptional and epigenetic reorganization that ultimately gives rise to the cellular and behavioral adaptations that drive addiction-linked behaviors. Within the NAc, DA engages the type-1 DA receptor (DRD1, encoded by the Drd1 gene) that is present on a subset of medium spiny neurons (MSNs), called D1-MSNs. DRD1 activation stimulates signal transduction cascades that results in CREB phosphorylation and activation of immediate early gene (IEG) expression programs. These IEG programs are comprised largely of activity-dependent transcription factors such as Fos, FosB, and Nr4a1 (Savell et al., 2020; Nishi et al., 2011; Bertran-Gonzalez et al., 2008), each of which are critical mediators for cellular and behavioral adaptations to drugs of abuse (Zhang et al., 2006; Nestler et al., 2001; Carpenter et al., 2020). The NAc is a heterogeneous brain region consisting of many transcriptionally defined cell types. While a significant amount of research has focused on the role of DA receptor-expressing MSNs in adaptations to drugs of abuse, recent reports have also demonstrated significant roles for other cell types. For example, optogenetic inhibition of NAc somatostatin expressing interneurons (Sst-interneurons) attenuates cocaine-induced conditioned place preference (Ribeiro et al., 2018), suggesting that these cells are important for behavioral adaptations to drugs of abuse. Similarly, photoinhibition of cholinergic interneurons (Chat-interneurons) also blocks cocaine-induced conditioned place preference (Witten et al., 2010). Interestingly, recent research has also demonstrated an important role for astrocytes in mediating cocaine-induced synaptic and behavioral adaptations (Wang et al., 2021; Corkrum et al., 2020). These results demonstrate that diverse cell types within the NAc contribute to the molecular, cellular, and behavioral effects of cocaine. Despite established cell type heterogeneity in the NAc, many studies rely on transgenic animal models that allow for the investigation of these changes only in individual cell populations. Unlike many techniques focused on a defined class of cells, single nucleus RNA-sequencing (snRNA-seq) provides unprecedented resolution that allows for the investigation of thousands of genes in thousands of cells with a single assay. This resolution not only permits detection of cocaine-induced transcriptional changes in many cell types, but also allows for better understanding of the molecular heterogeneity within the NAc. Recently, snRNA-seq approaches have been used to generate cellular atlases of the mouse (Chen et al., 2021), non-human primate (He et al., 2021), and human (Tran et al., 2021) NAc. However, snRNA-seq has only been used to define early gene expression changes to acute cocaine, while the transcriptional effects of repeated cocaine in various NAc cell types is less well studied. Here, we used snRNA-seq to transcriptionally profile 39,254 nuclei from male and female Sprague-Dawley rats. Rats were further subdivided by treatment (cocaine or saline) and exposure paradigm (acute or repeated). This study confirms both the presence of many previously defined cell types within the rat NAc, and reveals that D1-MSNs exhibit the most robust transcriptional response to cocaine across all cell populations. Additionally, snRNA-seq of the rat NAc allowed for the investigation of cluster-specific expression of different neurotransmitter receptor systems and identified novel marker genes for two populations of transcriptionally distinct D1-MSNs.\",\\n  \"entities\": [\\n    {\\n      \"entity\": \"Drug addiction\",\\n      \"start\": 0,\\n      \"end\": 13,\\n      \"label\": \"DISORDER\"\\n    },\\n    {\\n      \"entity\": \"neuropsychiatric disorder\",\\n      \"start\": 29,\\n      \"end\": 53,\\n      \"label\": \"DISORDER\"\\n    },\\n    {\\n      \"entity\": \"United States\",\\n      \"start\": 146,\\n      \"end\": 159,\\n      \"label\": \"LOCATION\"\\n    },\\n    {\\n      \"entity\": \"COVID-19 pandemic\",\\n      \"start\": 181,\\n      \"end\": 198,\\n      \"label\": \"EVENT\"\\n    },\\n    {\\n      \"entity\": \"Cisneros and Cunningham, 2021\",\\n      \"start\": 245,\\n      \"end\": 275,\\n      \"label\": \"CITATION\"\\n    },\\n    {\\n      \"entity\": \"opioids\",\\n      \"start\": 356,\\n      \"end\": 363,\\n      \"label\": \"DRUG_CLASS\"\\n    },\\n    {\\n      \"entity\": \"fentanyl\",\\n      \"start\": 395,\\n      \"end\": 403,\\n      \"label\": \"DRUG\"\\n    },\\n    {\\n      \"entity\": \"psychostimulants\",\\n      \"start\": 461,\\n      \"end\": 477,\\n      \"label\": \"DRUG_CLASS\"\\n    },\\n    {\\n      \"entity\": \"Hedegaard et al., 2021\",\\n      \"start\": 506,\\n      \"end\": 528,\\n      \"label\": \"CITATION\"\\n    },\\n    {\\n      \"entity\": \"cocaine\",\\n      \"start\": 584,\\n      \"end\": 591,\\n      \"label\": \"DRUG\"\\n    },\\n    {\\n      \"entity\": \"Hedegaard et al., 2021\",\\n      \"start\": 623,\\n      \"end\": 645,\\n      \"label\": \"CITATION\"\\n    },\\n    {\\n      \"entity\": \"Ciccarone, 2021\",\\n      \"start\": 647,\\n      \"end\": 662,\\n      \"label\": \"CITATION\"\\n    },\\n    {\\n      \"entity\": \"opioid use disorder\",\\n      \"start\": 688,\\n      \"end\": 706,\\n      \"label\": \"DISORDER\"\\n    },\\n    {\\n      \"entity\": \"stimulant use disorder\",\\n      \"start\": 745,\\n      \"end\": 766,\\n      \"label\": \"DISORDER\"\\n    },\\n    {\\n      \"entity\": \"mesolimbic dopamine pathway\",\\n      \"start\": 948,\\n      \"end\": 974,\\n      \"label\": \"NEURAL_PATHWAY\"\\n    },\\n    {\\n      \"entity\": \"ventral tegmental area\",\\n      \"start\": 1048,\\n      \"end\": 1071,\\n      \"label\": \"BRAIN_REGION\"\\n    },\\n    {\\n      \"entity\": \"VTA\",\\n      \"start\": 1073,\\n      \"end\": 1076,\\n      \"label\": \"BRAIN_REGION_ABBR\"\\n    },\\n    {\\n      \"entity\": \"nucleus accumbens\",\\n      \"start\": 1092,\\n      \"end\": 1108,\\n      \"label\": \"BRAIN_REGION\"\\n    },\\n    {\\n      \"entity\": \"NAc\",\\n      \"start\": 1110,\\n      \"end\": 1113,\\n      \"label\": \"BRAIN_REGION_ABBR\"\\n    },\\n    {\\n      \"entity\": \"dopamine\",\\n      \"start\": 1137,\\n      \"end\": 1145,\\n      \"label\": \"NEUROTRANSMITTER\"\\n    },\\n    {\\n      \"entity\": \"DA\",\\n      \"start\": 1147,\\n      \"end\": 1149,\\n      \"label\": \"NEUROTRANSMITTER_ABBR\"\\n    },\\n    {\\n      \"entity\": \"Day et al., 2007\",\\n      \"start\": 1219,\\n      \"end\": 1235,\\n      \"label\": \"CITATION\"\\n    },\\n    {\\n      \"entity\": \"cocaine\",\\n      \"start\": 1268,\\n      \"end\": 1275,\\n      \"label\": \"DRUG\"\\n    },\\n    {\\n      \"entity\": \"DA\",\\n      \"start\": 1313,\\n      \"end\": 1315,\\n      \"label\": \"NEUROTRANSMITTER_ABBR\"\\n    },\\n    {\\n      \"entity\": \"NAc\",\\n      \"start\": 1338,\\n      \"end\": 1341,\\n      \"label\": \"BRAIN_REGION_ABBR\"\\n    },\\n    {\\n      \"entity\": \"DA transporter\",\\n      \"start\": 1385,\\n      \"end\": 1398,\\n      \"label\": \"PROTEIN\"\\n    },\\n    {\\n      \"entity\": \"VTA\",\\n      \"start\": 1419,\\n      \"end\": 1422,\\n      \"label\": \"BRAIN_REGION_ABBR\"\\n    },\\n    {\\n      \"entity\": \"Giros et al., 1996\",\\n      \"start\": 1529,\\n      \"end\": 1547,\\n      \"label\": \"CITATION\"\\n    },\\n    {\\n      \"entity\": \"DA\",\\n      \"start\": 1568,\\n      \"end\": 1570,\\n      \"label\": \"NEUROTRANSMITTER_ABBR\"\\n    },\\n    {\\n      \"entity\": \"type-1 DA receptor\",\\n      \"start\": 1627,\\n      \"end\": 1644,\\n      \"label\": \"RECEPTOR\"\\n    },\\n    {\\n      \"entity\": \"DRD1\",\\n      \"start\": 1647,\\n      \"end\": 1651,\\n      \"label\": \"RECEPTOR\"\\n    },\\n    {\\n      \"entity\": \"Drd1\",\\n      \"start\": 1670,\\n      \"end\": 1674,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"medium spiny neurons\",\\n      \"start\": 1700,\\n      \"end\": 1719,\\n      \"label\": \"CELL_TYPE\"\\n    },\\n    {\\n      \"entity\": \"MSNs\",\\n      \"start\": 1721,\\n      \"end\": 1725,\\n      \"label\": \"CELL_TYPE_ABBR\"\\n    },\\n    {\\n      \"entity\": \"D1-MSNs\",\\n      \"start\": 1737,\\n      \"end\": 1744,\\n      \"label\": \"CELL_TYPE\"\\n    },\\n    {\\n      \"entity\": \"DRD1\",\\n      \"start\": 1746,\\n      \"end\": 1750,\\n      \"label\": \"RECEPTOR\"\\n    },\\n    {\\n      \"entity\": \"CREB\",\\n      \"start\": 1808,\\n      \"end\": 1812,\\n      \"label\": \"PROTEIN\"\\n    },\\n    {\\n      \"entity\": \"immediate early gene\",\\n      \"start\": 1834,\\n      \"end\": 1853,\\n      \"label\": \"GENE_CLASS\"\\n    },\\n    {\\n      \"entity\": \"IEG\",\\n      \"start\": 1855,\\n      \"end\": 1858,\\n      \"label\": \"GENE_CLASS_ABBR\"\\n    },\\n    {\\n      \"entity\": \"Fos\",\\n      \"start\": 1921,\\n      \"end\": 1924,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"FosB\",\\n      \"start\": 1926,\\n      \"end\": 1930,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"Nr4a1\",\\n      \"start\": 1932,\\n      \"end\": 1937,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"Savell et al., 2020\",\\n      \"start\": 1940,\\n      \"end\": 1960,\\n      \"label\": \"CITATION\"\\n    },\\n    {\\n      \"entity\": \"Nishi et al., 2011\",\\n      \"start\": 1962,\\n      \"end\": 1980,\\n      \"label\": \"CITATION\"\\n    },\\n    {\\n      \"entity\": \"Bertran-Gonzalez et al., 2008\",\\n      \"start\": 1982,\\n      \"end\": 2011,\\n      \"label\": \"CITATION\"\\n    },\\n    {\\n      \"entity\": \"Zhang et al., 2006\",\\n      \"start\": 2061,\\n      \"end\": 2080,\\n      \"label\": \"CITATION\"\\n    },\\n    {\\n      \"entity\": \"Nestler et al., 2001\",\\n      \"start\": 2082,\\n      \"end\": 2102,\\n      \"label\": \"CITATION\"\\n    },\\n    {\\n      \"entity\": \"Carpenter et al., 2020\",\\n      \"start\": 2104,\\n      \"end\": 2126,\\n      \"label\": \"CITATION\"\\n    },\\n    {\\n      \"entity\": \"NAc\",\\n      \"start\": 2131,\\n      \"end\": 2134,\\n      \"label\": \"BRAIN_REGION_ABBR\"\\n    },\\n    {\\n      \"entity\": \"somatostatin expressing interneurons\",\\n      \"start\": 2346,\\n      \"end\": 2382,\\n      \"label\": \"CELL_TYPE\"\\n    },\\n    {\\n      \"entity\": \"Sst-interneurons\",\\n      \"start\": 2384,\\n      \"end\": 2398,\\n      \"label\": \"CELL_TYPE_ABBR\"\\n    },\\n    {\\n      \"entity\": \"Ribeiro et al., 2018\",\\n      \"start\": 2451,\\n      \"end\": 2471,\\n      \"label\": \"CITATION\"\\n    },\\n    {\\n      \"entity\": \"cholinergic interneurons\",\\n      \"start\": 2512,\\n      \"end\": 2535,\\n      \"label\": \"CELL_TYPE\"\\n    },\\n    {\\n      \"entity\": \"Chat-interneurons\",\\n      \"start\": 2537,\\n      \"end\": 2552,\\n      \"label\": \"CELL_TYPE_ABBR\"\\n    },\\n    {\\n      \"entity\": \"Witten et al., 2010\",\\n      \"start\": 2607,\\n      \"end\": 2626,\\n      \"label\": \"CITATION\"\\n    },\\n    {\\n      \"entity\": \"astrocytes\",\\n      \"start\": 2679,\\n      \"end\": 2689,\\n      \"label\": \"CELL_TYPE\"\\n    },\\n    {\\n      \"entity\": \"Wang et al., 2021\",\\n      \"start\": 2761,\\n      \"end\": 2779,\\n      \"label\": \"CITATION\"\\n    },\\n    {\\n      \"entity\": \"Corkrum et al., 2020\",\\n      \"start\": 2781,\\n      \"end\": 2801,\\n      \"label\": \"CITATION\"\\n    },\\n    {\\n      \"entity\": \"NAc\",\\n      \"start\": 2816,\\n      \"end\": 2819,\\n      \"label\": \"BRAIN_REGION_ABBR\"\\n    },\\n    {\\n      \"entity\": \"single nucleus RNA-sequencing\",\\n      \"start\": 3076,\\n      \"end\": 3104,\\n      \"label\": \"TECHNIQUE\"\\n    },\\n    {\\n      \"entity\": \"snRNA-seq\",\\n      \"start\": 3106,\\n      \"end\": 3115,\\n      \"label\": \"TECHNIQUE_ABBR\"\\n    },\\n    {\\n      \"entity\": \"NAc\",\\n      \"start\": 3289,\\n      \"end\": 3292,\\n      \"label\": \"BRAIN_REGION_ABBR\"\\n    },\\n    {\\n      \"entity\": \"mouse\",\\n      \"start\": 3433,\\n      \"end\": 3438,\\n      \"label\": \"ORGANISM\"\\n    },\\n    {\\n      \"entity\": \"Chen et al., 2021\",\\n      \"start\": 3441,\\n      \"end\": 3459,\\n      \"label\": \"CITATION\"\\n    },\\n    {\\n      \"entity\": \"non-human primate\",\\n      \"start\": 3462,\\n      \"end\": 3478,\\n      \"label\": \"ORGANISM\"\\n    },\\n    {\\n      \"entity\": \"He et al., 2021\",\\n      \"start\": 3483,\\n      \"end\": 3498,\\n      \"label\": \"CITATION\"\\n    },\\n    {\\n      \"entity\": \"human\",\\n      \"start\": 3503,\\n      \"end\": 3508,\\n      \"label\": \"ORGANISM\"\\n    },\\n    {\\n      \"entity\": \"Tran et al., 2021\",\\n      \"start\": 3511,\\n      \"end\": 3529,\\n      \"label\": \"CITATION\"\\n    },\\n    {\\n      \"entity\": \"NAc\",\\n      \"start\": 3534,\\n      \"end\": 3537,\\n      \"label\": \"BRAIN_REGION_ABBR\"\\n    },\\n    {\\n      \"entity\": \"Sprague-Dawley rats\",\\n      \"start\": 3712,\\n      \"end\": 3730,\\n      \"label\": \"ORGANISM\"\\n    },\\n    {\\n      \"entity\": \"cocaine\",\\n      \"start\": 3757,\\n      \"end\": 3764,\\n      \"label\": \"DRUG\"\\n    },\\n    {\\n      \"entity\": \"saline\",\\n      \"start\": 3769,\\n      \"end\": 3775,\\n      \"label\": \"DRUG\"\\n    },\\n    {\\n      \"entity\": \"NAc\",\\n      \"start\": 3910,\\n      \"end\": 3913,\\n      \"label\": \"BRAIN_REGION_ABBR\"\\n    },\\n    {\\n      \"entity\": \"D1-MSNs\",\\n      \"start\": 3981,\\n      \"end\": 3988,\\n      \"label\": \"CELL_TYPE\"\\n    },\\n    {\\n      \"entity\": \"NAc\",\\n      \"start\": 4110,\\n      \"end\": 4113,\\n      \"label\": \"BRAIN_REGION', additional_kwargs={'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 4096, 'prompt_tokens': 1211, 'total_tokens': 5307, 'completion_tokens_details': None, 'prompt_tokens_details': None}, 'model_name': 'deepseek/deepseek-chat-v3-0324:free', 'system_fingerprint': None, 'id': 'gen-1746228846-2DQyFmPPdCMBrBsIFYlM', 'finish_reason': 'length', 'logprobs': None}, id='run-60a7cfa1-b8e7-4fed-925d-345060b18009-0', usage_metadata={'input_tokens': 1211, 'output_tokens': 4096, 'total_tokens': 5307, 'input_token_details': {}, 'output_token_details': {}}),\n",
       " AIMessage(content='```# 1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1极简主义\\n\\n极简主义是一种生活哲学和艺术风格，强调通过减少不必要的元素，保留最核心、最本质的内容，以达到简单、清晰和高效的目的。在艺术、设计、建筑、音乐、文学以及生活方式等多个领域，极简主义都有广泛的应用和影响。\\n\\n### 极简主义的特点：\\n1. **简单性**：去除多余的装饰和复杂的设计，追求简洁明了的形式。\\n2. **功能性**：注重实用性和功能性，每一个元素都有其存在的理由。\\n3. **留白**：通过留白和空间的使用，增强视觉或感官上的平衡与和谐。\\n4. **自然与真实**：倾向于使用天然材料，展现材料的本质，避免过度加工。\\n5. **情感与专注**：通过简化外在的干扰，帮助人们专注于真正重要的事物，提升内在的平静与满足感。\\n\\n### 极简主义在不同领域的体现：\\n- **设计**：极简设计强调“少即是多”（Less is more），如苹果公司的产品设计。\\n- **建筑**：极简主义建筑通常使用简单的几何形状、中性色调和开放空间，如日本建筑师安藤忠雄的作品。\\n- **生活方式**：极简生活主张减少物质拥有，只保留真正需要和喜爱的物品，避免过度消费和浪费。\\n- **艺术**：极简主义艺术以简单的几何形状和单色为主，如唐纳德·贾德（Donald Judd）的雕塑。\\n\\n### 极简主义的优点：\\n- **减少压力**：通过简化生活，减少杂乱和混乱，从而降低心理压力。\\n- **提高效率**：在工作和生活中，极简主义可以帮助人们更高效地完成任务。\\n- **环保**：减少不必要的消费和浪费，更符合可持续发展的理念。\\n\\n### 如何实践极简主义：\\n1. **物品**：定期清理家中不必要的物品，只保留真正有用或带来快乐的东西。\\n2. **时间**：简化日程，优先处理重要的事情，避免过度忙碌。\\n3. **数字生活**：清理电子设备中不必要的文件和应用，减少信息过载。\\n4. **消费**：购买时三思而行，选择高质量、耐用的物品，避免冲动消费。\\n\\n极简主义不仅仅是一种风格，更是一种生活态度，它鼓励人们通过简化外在的复杂性，找到内在的平衡与自由。', additional_kwargs={'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 2702, 'prompt_tokens': 769, 'total_tokens': 3471, 'completion_tokens_details': None, 'prompt_tokens_details': None}, 'model_name': 'deepseek/deepseek-chat-v3-0324:free', 'system_fingerprint': None, 'id': 'gen-1746228955-LE5zA3AjrTqhmx7qXEQu', 'finish_reason': 'stop', 'logprobs': None}, id='run-48c97f7f-392d-4f24-ae05-bbdbcf830384-0', usage_metadata={'input_tokens': 769, 'output_tokens': 2702, 'total_tokens': 3471, 'input_token_details': {}, 'output_token_details': {}}),\n",
       " AIMessage(content='Here is the training-ready JSON object for spaCy NER training with neuroscience-specific entity labels:\\n\\n```json\\n{\\n  \"text\": \"The integration of over 39,000 nuclei from rats that underwent different cocaine exposure paradigms identified two populations of Drd1-expressing MSNs (Drd1-MSN-1 and Drd1-MSN-2; Fig. 2a). To begin understanding the distinct gene expression signatures of these Drd1-MSN populations, we conducted a differential expression analysis using pseudobulked gene expression matrices. This analysis identified 506 genes enriched in the Drd1-MSN-1 cluster and 665 genes enriched in the Drd1-MSN-2 cluster. Interestingly, Ebf1 is the most highly enriched differentially expressed gene in the Drd1-MSN-1 population (Fig. 2b). Ebf1 is a known marker of D1-MSNs and is implicated in MSN early development (Lobo et al., 2008). Calb1, the gene encoding the calcium binding protein Calbindin 1, is another gene that is enriched in the Drd1-MSN-1 population (Fig. 2b). Calb1 is a spatial marker within the NAc and is highly enriched in the NAc core (Jongen-Rêlo et al., 1994). Included in the top 10 differentially expressed genes in the Drd1-MSN-2 population was Htr4, the gene that codes for the serotonin receptor 4 (Fig. 2b). HTR4 autoradiography suggests that this receptor is located in the ventral portion of the NAc shell (Jean et al., 2012). The differential enrichment of Htr4 and Calb1 suggests that the two populations of Drd1expressing MSNs identified by snRNA-seq are both transcriptionally and spatially distinct. Next, we set out to understand the specific gene networks and pathways that may be enriched in these two D1-MSN populations. Molecular function pathway analysis revealed an enrichment of genes involved in calcium transport and signaling in the Drd1-MSN-1 population, but enrichment of genes involved in potassium transport in the Drd1-MSN-2 population (Fig. 2c). For example, Trpc3, a gene encoding a calcium permeable channel, and Cacng3, a gene encoding an L-type calcium channel, are both highly enriched in the Drd1-MSN-1 population (Fig. 2d). Conversely, several genes encoding potassium channels, such as Kcnc1, Kcnc2, and Kcng3, are enriched in the Drd1-MSN-2 population, further highlighting the enrichment of genes involved in potassium transport in this population. While these genes are differentially enriched, both calcium-and potassium-related genes are expressed to some extent in both clusters (Fig. 2d). Functional investigation of transcriptionally distinct populations of neurons in the mammalian brain requires the use of cell type-specific promoters. To facilitate the future investigation of the two populations of transcriptionally distinct Drd1-expressing MSNs, we next identified potential marker genes that are both differentially and highly expressed within a population. This analysis identified Ebf1 as a potential marker gene for the Drd1-MSN-1 population (Fig. 2e) and Htr4 as a potential marker gene for the Drd1-MSN-2 population (Fig. 2f). An additional characteristic of marker genes is that they are expressed in a large proportion of cells within the cluster. Plotting the expression of these genes on a UMAP shows differential enrichment of Ebf1 and Htr4 between the Drd1-expressing MSN populations (Fig. 2g). However, Ebf1 is also expressed in endothelial cells, while Htr4 is also expressed in Drd2and Drd3-MSNs. Thus, these genes may act as marker genes in combination with Drd1 but cannot act as marker genes alone.\",\\n  \"entities\": [\\n    {\\n      \"entity\": \"cocaine\",\\n      \"start\": 80,\\n      \"end\": 87,\\n      \"label\": \"SUBSTANCE\"\\n    },\\n    {\\n      \"entity\": \"Drd1-expressing MSNs\",\\n      \"start\": 150,\\n      \"end\": 171,\\n      \"label\": \"CELL_TYPE\"\\n    },\\n    {\\n      \"entity\": \"Drd1-MSN-1\",\\n      \"start\": 174,\\n      \"end\": 184,\\n      \"label\": \"CELL_TYPE\"\\n    },\\n    {\\n      \"entity\": \"Drd1-MSN-2\",\\n      \"start\": 189,\\n      \"end\": 199,\\n      \"label\": \"CELL_TYPE\"\\n    },\\n    {\\n      \"entity\": \"Ebf1\",\\n      \"start\": 380,\\n      \"end\": 384,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"D1-MSNs\",\\n      \"start\": 420,\\n      \"end\": 427,\\n      \"label\": \"CELL_TYPE\"\\n    },\\n    {\\n      \"entity\": \"Calb1\",\\n      \"start\": 493,\\n      \"end\": 498,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"Calbindin 1\",\\n      \"start\": 528,\\n      \"end\": 539,\\n      \"label\": \"PROTEIN\"\\n    },\\n    {\\n      \"entity\": \"NAc\",\\n      \"start\": 570,\\n      \"end\": 573,\\n      \"label\": \"BRAIN_REGION\"\\n    },\\n    {\\n      \"entity\": \"NAc core\",\\n      \"start\": 600,\\n      \"end\": 608,\\n      \"label\": \"BRAIN_REGION\"\\n    },\\n    {\\n      \"entity\": \"Htr4\",\\n      \"start\": 700,\\n      \"end\": 704,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"serotonin receptor 4\",\\n      \"start\": 730,\\n      \"end\": 751,\\n      \"label\": \"PROTEIN\"\\n    },\\n    {\\n      \"entity\": \"NAc shell\",\\n      \"start\": 840,\\n      \"end\": 849,\\n      \"label\": \"BRAIN_REGION\"\\n    },\\n    {\\n      \"entity\": \"Trpc3\",\\n      \"start\": 1200,\\n      \"end\": 1205,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"Cacng3\",\\n      \"start\": 1265,\\n      \"end\": 1271,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"Kcnc1\",\\n      \"start\": 1400,\\n      \"end\": 1405,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"Kcnc2\",\\n      \"start\": 1407,\\n      \"end\": 1412,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"Kcng3\",\\n      \"start\": 1414,\\n      \"end\": 1419,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"Drd2\",\\n      \"start\": 2200,\\n      \"end\": 2204,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"Drd3\",\\n      \"start\": 2205,\\n      \"end\": 2209,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"Drd1\",\\n      \"start\": 2226,\\n      \"end\": 2230,\\n      \"label\": \"GENE\"\\n    }\\n  ]\\n}\\n```\\n\\n### Entity Label Definitions:\\n1. **SUBSTANCE**: Psychoactive or pharmacological substances (e.g., cocaine).\\n2. **CELL_TYPE**: Specific neuronal or cellular populations (e.g., Drd1-MSN-1, D1-MSNs).\\n3. **GENE**: Gene symbols or names (e.g., Ebf1, Htr4, Trpc3).\\n4. **PROTEIN**: Protein names (e.g., Calbindin 1, serotonin receptor 4).\\n5. **BRAIN_REGION**: Neuroanatomical regions (e.g., NAc, NAc core, NAc shell).\\n\\nThe output is structured for direct use in spaCy\\'s NER training pipeline, with precise neuroscience domain labels. Let me know if you\\'d like any refinements!', additional_kwargs={'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 1773, 'prompt_tokens': 904, 'total_tokens': 2677, 'completion_tokens_details': None, 'prompt_tokens_details': None}, 'model_name': 'deepseek/deepseek-chat-v3-0324:free', 'system_fingerprint': None, 'id': 'gen-1746229065-tZZbaeJ9HnMwdCdLSMvC', 'finish_reason': 'stop', 'logprobs': None}, id='run-3f5a4933-257c-4b9d-a5ac-43639b63e532-0', usage_metadata={'input_tokens': 904, 'output_tokens': 1773, 'total_tokens': 2677, 'input_token_details': {}, 'output_token_details': {}}),\n",
       " AIMessage(content='', additional_kwargs={'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 1, 'prompt_tokens': 383, 'total_tokens': 384, 'completion_tokens_details': None, 'prompt_tokens_details': None}, 'model_name': 'deepseek/deepseek-chat-v3-0324:free', 'system_fingerprint': None, 'id': 'gen-1746229110-qcH8G7cNA2v8ddibL8v0', 'finish_reason': 'stop', 'logprobs': None}, id='run-4a6c9ae2-3ea1-4aab-a69e-93be2c99ecaf-0', usage_metadata={'input_tokens': 383, 'output_tokens': 1, 'total_tokens': 384, 'input_token_details': {}, 'output_token_details': {}}),\n",
       " AIMessage(content='Here is the formatted JSON object for spaCy NER training, focusing on neuroscience-specific entities:\\n\\n```json\\n{\\n  \"text\": \"We previously demonstrated that a single acute cocaine exposure differentially alters the transcriptome of molecularly defined cell types within the NAc (Savell et al., 2020). To begin understanding how repeated cocaine exposure alters the transcriptome of distinct NAc population, we injected male and female Sprague-Dawley rat with 20 mg/kg cocaine for 7 consecutive days (Fig. 1a, Supp. Fig. 1). Before directly comparing acute and repeated exposures, we first investigated cell type-specific transcriptional response to cocaine exposure, regardless of type or duration. To do this, we collapsed both datasets and used a pseudobulk approach to identify cocaine-dependent, cell type-specific differentially expressed gene (DEG; Supp. Table 2). This approach increased the number of replicates used for statistical comparison and circumvented previously identified issues with deprecated differential expression method previously used for snRNA-seq (Squair et al., 2021). Strikingly, the Drd1-MSN-1 population exhibited the most robust transcriptional response to cocaine, with 236 total DEGs following cocaine experience (184 upregulated, 52 downregulated; Fig. 4a,b). These upregulated DEGs include immediate early gene (IEG) such as Nr4a3, Tiparp, and Fosl2 (Fig. 4b). Recent evidence has demonstrated that astrocyte within the NAc also play a key role in the neuronal and behavioral response to dopamine and drug of abuse (Wang et al., 2021; Corkrum et al., 2020; Shelkar et al., 2022). Disruption of glutamatergic transport, release, and signaling is implicated in substance use disorder (Reissner and Kalivas, 2010; Kalivas, 2009), and astrocyte are heavily involved in the regulation of this process across the brain (Mahmoud et al., 2019). Interestingly, Gria1, an ionotropic AMPA receptor, is significantly decreased in astrocyte following cocaine exposure (Fig. 4c). Unlike the Drd1-MSN-1 population, astrocyte exhibited more gene with cocaine-dependent decrease in expression (Fig. 4c,g). While most DEGs in the astrocyte population are downregulated with cocaine, Pde10a, the gene encoding the cAMP-regulated phosphodiesterase 10A, is upregulated by cocaine in both the astrocyte and Sst-interneuron cluster (Fig. 4c,d). Previous research has demonstrated that Drd1-expressing MSN exhibit robust transcriptional change in response to cocaine (Savell et al., 2020; Bertran-Gonzalez et al., 2008; Zhang et al., 2006; Zhang et al., 2005). More specifically, snRNA-seq from the rat NAc following acute cocaine exposure demonstrated that D1-MSN have the most robust transcriptional response to cocaine when compared to other transcriptionally distinct NAc cell type (Savell et al., 2020). In this dataset, only one of the Drd1-expressing MSN population (Drd1-MSN-1) had a robust transcriptional response to cocaine. For example, Nr4a3, Fosb, Fosl2, and Tiparp are all significantly increased following cocaine exposure only in the Drd1-MSN-1 population (Fig. 4e). Next, we set out to investigate the relationship of cocaine-dependent DEG across cocaine exposure paradigm. To do this, we first conducted DEG analysis comparing cocaine and saline sample with exposure type ignored (i.e., all cocaine sample vs all saline sample), acute exposure only, and repeated exposure sample only. The log2 (Fold Change) value for all DEG was then concatenated and subject to hierarchical clustering. Overall, this analysis identified that many of the gene altered by acute cocaine are also altered by repeated cocaine in the same direction (Fig. 4f-h). However, for the Drd1-MSN-1 population some gene altered by acute cocaine are not altered by repeated cocaine. Furthermore, while the overall pattern shows concordant gene expression pattern, some cocaine-regulated gene within the astrocyte cluster demonstrate opposite change with an acute or repeated exposure. To estimate the degree of concordance between gene expression program altered by acute and repeated cocaine exposure in a threshold-free manner, we next used the rank-rank hypergeometric overlap (RRHO) algorithm (Cahill et al., 2018). To complete this analysis, we ranked gene list by the log2 (Fold Change), which was calculated by dividing the log-normalized expression value of cocaine cell by the log-normalized expression value of saline cell. This analysis identified that, on a cell type-specific basis, acute and repeated cocaine gene expression signature are primarily concordant. For example, the acute and repeated cocaine change for the Drd1-MSN-1, astrocyte, and Sst-interneuron population are significantly concordant for both upregulated and downregulated gene (Fig. 4i-l). However, the strength of this overlap and concordance is substantially different for these population.\",\\n  \"entities\": [\\n    {\\n      \"entity\": \"NAc\",\\n      \"start\": 120,\\n      \"end\": 123,\\n      \"label\": \"BRAIN_REGION\"\\n    },\\n    {\\n      \"entity\": \"Sprague-Dawley rat\",\\n      \"start\": 250,\\n      \"end\": 268,\\n      \"label\": \"ORGANISM\"\\n    },\\n    {\\n      \"entity\": \"Drd1-MSN-1\",\\n      \"start\": 800,\\n      \"end\": 810,\\n      \"label\": \"NEURON_TYPE\"\\n    },\\n    {\\n      \"entity\": \"Nr4a3\",\\n      \"start\": 1000,\\n      \"end\": 1005,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"Tiparp\",\\n      \"start\": 1007,\\n      \"end\": 1013,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"Fosl2\",\\n      \"start\": 1015,\\n      \"end\": 1020,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"astrocyte\",\\n      \"start\": 1200,\\n      \"end\": 1209,\\n      \"label\": \"CELL_TYPE\"\\n    },\\n    {\\n      \"entity\": \"Gria1\",\\n      \"start\": 1500,\\n      \"end\": 1505,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"Pde10a\",\\n      \"start\": 1800,\\n      \"end\": 1806,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"Sst-interneuron\",\\n      \"start\": 1900,\\n      \"end\": 1914,\\n      \"label\": \"NEURON_TYPE\"\\n    },\\n    {\\n      \"entity\": \"D1-MSN\",\\n      \"start\": 2200,\\n      \"end\": 2206,\\n      \"label\": \"NEURON_TYPE\"\\n    },\\n    {\\n      \"entity\": \"Fosb\",\\n      \"start\": 2500,\\n      \"end\": 2504,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"Fosl2\",\\n      \"start\": 2506,\\n      \"end\": 2511,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"Tiparp\",\\n      \"start\": 2513,\\n      \"end\": 2519,\\n      \"label\": \"GENE\"\\n    }\\n  ]\\n}\\n```\\n\\n### Key Features:\\n1. **Entity Labels**:\\n   - `BRAIN_REGION` (e.g., NAc)\\n   - `ORGANISM` (e.g., Sprague-Dawley rat)\\n   - `NEURON_TYPE` (e.g., Drd1-MSN-1, Sst-interneuron)\\n   - `CELL_TYPE` (e.g., astrocyte)\\n   - `GENE` (e.g., Nr4a3, Gria1)\\n\\n2. **Format Compliance**:\\n   - Follows the spaCy NER training format with `entity`, `start`, `end`, and `label`.\\n   - Covers neuroscience-specific terms (genes, cell types, brain regions).\\n\\n3. **Contextual Precision**:\\n   - Labels are applied to terms in their relevant biological context (e.g., \"Drd1-MSN-1\" as a neuron type, not a gene).\\n\\n4. **Scalability**:\\n   - The JSON structure can be extended to include additional entities or text samples.\\n\\nLet me know if you\\'d like to refine the labels or add more entities!', additional_kwargs={'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 1836, 'prompt_tokens': 1151, 'total_tokens': 2987, 'completion_tokens_details': None, 'prompt_tokens_details': None}, 'model_name': 'deepseek/deepseek-chat-v3-0324:free', 'system_fingerprint': None, 'id': 'gen-1746229112-XH1S3YM4bEwuSd2pwAhE', 'finish_reason': 'stop', 'logprobs': None}, id='run-d9dd09d7-a7a7-4240-b0f7-2145877518f2-0', usage_metadata={'input_tokens': 1151, 'output_tokens': 1836, 'total_tokens': 2987, 'input_token_details': {}, 'output_token_details': {}}),\n",
       " AIMessage(content='{\\n  \"text\": \"Recently, transcriptional atlases of the non-human primate (NHP) striatum (He et al., 2021) and human NAc (Tran et al., 2021) have demonstrated similar gene expression profiles of specific marker genes that we identify within the rat NAc. To quantify this overlap and perform cross-species comparisons, we focused on conservation of MSNs given that both NHP and human datasets identified and validated the presence of several transcriptionally distinct MSN subtypes. Direct comparison of our rat snRNA-seq dataset to monkey or human datasets revealed potential conservation of many NHP and human cell types (Fig. 5a-c). For example, the rat Drd1-MSN-1 population is correlated with the human D1.MSN_A and NHP D1.Matrix populations, and the rat Drd2-MSN-1 population is correlated with the human D2.MSN_A and the NHP D2. Matrix population (Fig. 5a-d). Transcriptionally conserved cell types should exhibit similar expression and enrichment of marker genes. For example, Drd2-MSNs are classically characterized by the enrichment of Drd2 and Penk, while Drd1-MSNs are characterized by the enrichment of Drd1 and Ebf1. Across species, transcriptionally distinct populations of Drd2-MSNs are similarly enriched for Drd2, but differ in their enrichment of Penk (Fig. 5c, Supp. Fig. 4a-c). A similar pattern is observed in Drd1-MSNs across species, as Drd1 is highly enriched in Drd1-MSNs, but Ebf1 enrichment differs (Fig. 5c, Supp. Fig. 4a-c). While the differential expression of Htr4 or Ebf1 distinguishes Drd1-MSNs in the rat NAc, this pattern is not observed in higher order mammals as Htr4 is ubiquitously expressed across cell types (Supp. Fig. 4a-c). Furthermore, Htr4 is enriched in the MSN.D1_A human population (Fig. 5c) along with Ebf1. Thus, Htr4 and Ebf1 may not be suitable markers for distinguishing between transcriptionally distinct populations of Drd1-MSNs in higher order mammals.\",\\n  \"entities\": [\\n    {\\n      \"entity\": \"non-human primate\",\\n      \"start\": 38,\\n      \"end\": 56,\\n      \"label\": \"ANIMAL_MODEL\"\\n    },\\n    {\\n      \"entity\": \"NHP\",\\n      \"start\": 57,\\n      \"end\": 60,\\n      \"label\": \"ABBREVIATION\"\\n    },\\n    {\\n      \"entity\": \"human NAc\",\\n      \"start\": 64,\\n      \"end\": 75,\\n      \"label\": \"BRAIN_REGION\"\\n    },\\n    {\\n      \"entity\": \"NAc\",\\n      \"start\": 76,\\n      \"end\": 79,\\n      \"label\": \"BRAIN_REGION\"\\n    },\\n    {\\n      \"entity\": \"MSNs\",\\n      \"start\": 188,\\n      \"end\": 193,\\n      \"label\": \"NEURON_TYPE\"\\n    },\\n    {\\n      \"entity\": \"Drd1-MSN-1\",\\n      \"start\": 304,\\n      \"end\": 317,\\n      \"label\": \"NEURON_SUBTYPE\"\\n    },\\n    {\\n      \"entity\": \"human D1.MSN_A\",\\n      \"start\": 370,\\n      \"end\": 386,\\n      \"label\": \"NEURON_SUBTYPE\"\\n    },\\n    {\\n      \"entity\": \"NHP D1.Matrix\",\\n      \"start\": 394,\\n      \"end\": 409,\\n      \"label\": \"NEURON_SUBTYPE\"\\n    },\\n    {\\n      \"entity\": \"Drd2-MSN-1\",\\n      \"start\": 422,\\n      \"end\": 435,\\n      \"label\": \"NEURON_SUBTYPE\"\\n    },\\n    {\\n      \"entity\": \"human D2.MSN_A\",\\n      \"start\": 464,\\n      \"end\": 480,\\n      \"label\": \"NEURON_SUBTYPE\"\\n    },\\n    {\\n      \"entity\": \"NHP D2.Matrix\",\\n      \"start\": 486,\\n      \"end\": 502,\\n      \"label\": \"NEURON_SUBTYPE\"\\n    },\\n    {\\n      \"entity\": \"Drd2\",\\n      \"start\": 533,\\n      \"end\": 537,\\n      \"label\": \"RECEPTOR_GENE\"\\n    },\\n    {\\n      \"entity\": \"Penk\",\\n      \"start\": 542,\\n      \"end\": 547,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"Drd1\",\\n      \"start\": 563,\\n      \"end\": 568,\\n      \"label\": \"RECEPTOR_GENE\"\\n    },\\n    {\\n      \"entity\": \"Ebf1\",\\n      \"start\": 573,\\n      \"end\": 578,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"Htr4\",\\n      \"start\": 607,\\n      \"end\": 612,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"MSN.D1_A\",\\n      \"start\": 629,\\n      \"end\": 638,\\n      \"label\": \"NEURON_POPULATION\"\\n    }\\n  ]\\n}', additional_kwargs={'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 1139, 'prompt_tokens': 583, 'total_tokens': 1722, 'completion_tokens_details': None, 'prompt_tokens_details': None}, 'model_name': 'deepseek/deepseek-chat-v3-0324:free', 'system_fingerprint': None, 'id': 'gen-1746229230-b4MdTCOjyX6xNj9pT8Qs', 'finish_reason': 'stop', 'logprobs': None}, id='run-10a0ff97-27e3-433e-b8b8-c45c25644ef5-0', usage_metadata={'input_tokens': 583, 'output_tokens': 1139, 'total_tokens': 1722, 'input_token_details': {}, 'output_token_details': {}}),\n",
       " AIMessage(content='Here is the training-ready JSON object for spaCy NER training with neuroscience-specific entity labels:\\n\\n```json\\n{\\n  \"text\": \"Within the past few decades, a major focus of drug addiction research has been on the transcriptional and epigenetic changes occurring in the NAc following drug exposure. While these studies have been integral in our understanding of how drug experience shapes the brain, they have Fig. 5. Identification of conserved neuronal cell types across higher order mammals. a-b, Heatmap colored by the Pearsons correlation coefficient between rat, nonhuman primate, and human cell types. Correlation coefficients were calculated using t-statistic for rat MSN DEGs and their orthologs. c, Heatmaps colored by tstatistic for specific MSN marker genes in rat, non-human primate, and human. d, Estimated conservation of NAc neuronal cell types across rat, non-human primate, and human. primarily used techniques that ignore cell-type heterogeneity. Furthermore, although the NAc is fundamental in reward processing and the rat is a primary model organism used for reward-related research, few studies have comprehensively profiled the rat NAc at the transcriptional level. snRNA-seq provides the unique opportunity to study both the cell type-specific transcriptional responses to drugs of abuse, as well as the cell-type heterogeneity that exists within the NAc. In the present study, we report the integration of a previously published NAc snRNA-seq dataset collected following acute cocaine exposure (Savell et al., 2020) with a new dataset collected following repeated cocaine exposure. The resulting dataset contains 39,254 nuclei across 16 transcriptionally distinct populations (11 neuronal and 5 non-neuronal), which more than doubles the size of our previously published work. The addition of both biological and technical replicates, along with the addition of over 24,000 nuclei, allowed us to generate a comprehensive transcriptional atlas of the rat NAc and investigate cell type-specific transcriptional responses to cocaine. We were also interested in understanding the distribution of expression of genes encoding different receptor systems, which provides additional information regarding which cell types may respond to diverse classes of drugs. For example, there is significant renewed interest in the action of psychedelic drugs, many of which act on serotonin receptors. Htr2a, the gene encoding serotonin receptor type 2A and the expected target of many psychedelics, is expressed in a small proportion of Drd1-and Drd2-expressing MSNs, as well as Sst-interneurons (Supp. Fig. 5a). Thus, these populations of cells will most likely respond to classic psychedelics. There is also considerable interest in the distribution of expression of opioid receptors in the NAc as opioid-associated overdose deaths continue to increase (Hedegaard et al., 2021;Ciccarone, 2021). Previous research has demonstrated that 25  of cells containing the delta opioid receptor (Oprd1) in the NAc also contained Met 5 -Enkaphalin (Svingos et al., 1998), a peptide encoded by the gene Penk. Penk is a marker gene for Drd2-expressing MSNs, suggesting that the delta opioid receptor is expressed in this population of neurons. Our data confirms this observation, as Oprd1 is highly expressed in the Drd2-MSN-1 population (Supp. Fig. 5a-c). A major focus of this study was to characterize cell type-specific gene programs induced by cocaine experience. Prior work has revealed distinct differences in transcriptomic changes between volitional cocaine paradigms, such has self-administration, and experimenteradministered injections. For example, after a prolonged withdrawal period, animals that underwent cocaine self-administration exhibit increased surface expression of calcium permeable AMPA receptors, while animals that underwent experimenter-administered injections do not (McCutcheon et al., 2011). Additionally, a recent study systematically tested transcriptomic changes occurring in the VTA across several cocaine paradigms and identified several IEGs that are upregulated by experimenter-administered cocaine injections (Campbell et al., 2021). Like this study, we also identified several IEGs (Fosl2, Tiparp, Fosb), that are upregulated by experimenter-administered cocaine injections. Notably, we did not observe significant locomotor sensitization to repeated doses of cocaine in the present study (Supp. Fig. 1b). While sensitization is commonly (but not always (Gulley et al., 2003;Mayfield et al., 1992;Cass et al., 1993)) observed with repeated cocaine injections, it is possible that distinct gene expression patterns may be observed in animals that exhibit sensitized responses to cocaine as compared to those that do not. This will be an interesting consideration for future studies that are able to obtain single-nuclei transcriptional profiles while preserving individual animal variability in the data structure. Ultimately, the results presented in the current study shed light on cocaine-induced, cell-type specific transcriptomic changes occurring in the NAc, expanding our understanding of how drug experience shapes the brain. Further research investigating paradigmspecific transcriptional changes may identify molecular correlates of more complex behaviors, such as cocaine seeking or craving. The largest transcriptional response to cocaine experience occurred in the Drd1-MSN-1 population (Fig. 4a,b). These changes included IEGs such as Fosl2. IEGs and other cocaine-regulated genes in the Drd1-MSN-1 population are induced by both acute and repeated cocaine, but to a lesser extent with repeated cocaine. This data is in agreement with previous research that demonstrated induction of FOSB protein by repeated cocaine was weaker as compared to acute cocaine (Zhang et al., 2005). The second population of Drd1-expressing MSNs within the NAc, Drd1-MSN-2, did not have a robust transcriptional response to cocaine experience. A multitude of previous research has demonstrated that Drd1-expressing cells within the NAc exhibit transcriptional and epigenetic responses to drugs of abuse (Savell et al., 2020;Bertran-Gonzalez et al., 2008;Zhang et al., 2006;Zhang et al., 2005;Zipperly et al., 2021). Thus, the identification of a population of Drd1-expressing MSNs in the NAc with a significantly attenuated transcriptional response to cocaine is intriguing because it highlights a previously underappreciated level of molecular heterogeneity within this population of cells. Notably, further analysis of population-specific DEGs also revealed an enrichment of calcium transport-related terms for Drd1-MSN-1 and potassium transport-related terms for Drd1-MSN-2. The differential enrichment for calcium-related terms and genes may explain the transcriptional response to cocaine in the Drd1-MSN-1 population, as many IEGs rely on calcium-dependent signaling cascades for induction. Our dataset also provided the opportunity to identify marker genes for these two populations. We identified Ebf1 as a marker gene for the Drd1-MSN-1 population and Htr4 as a marker gene for the Drd1-MSN-2 population. In situ experiments validated the presence of two distinct populations of Drd1-expressing MSNs based on the presence of Htr4 and Ebf1. These in situ experiments also revealed a spatial separation of these MSNs, as Drd1Ebf1 cells were primarily found in the NAc core and Drd1Htr4 cells were primarily found in the NAc shell. Transcriptional similarity based on spatial location may also explain why the Drd1-MSN-1 population is more closely related to the Drd2-MSN-1 population than the Drd1-MSN-2 population (Supp. Fig. 4a). In fact, analysis of the relationship between cell types within the NHP identifies that Drd1-and Drd2-expressing MSNs from the same striatal subregions are more closely related than Drd1-expressing MSNs found in distinct striatal compartments (Supp. Fig. 4b). Similarly, the MSN.D1_A and MSN.D2_A human populations are more closely related to each other than other Drd1-or Drd2-expressing MSNs, although no spatial information is available in that dataset (Supp. Fig. 4c). Together, these results suggest that at least some component of transcriptional diversity in the NAc reflects the spatial organization of this structure. Although sex differences in drug-induced behavioral adaptations have long been appreciated (Hu et al., 2004;Bobzean et al., 2014;Becker and Chartoff, 2019;Evans, 2007), more recent research has also demonstrated sex-specific drug-induced molecular adaptations (López et al., 2021). Our dataset includes nuclei from both male and female rats, and thus we examined sex-specific transcriptional responses to cocaine across cell populations within the NAc, again using the collapsed acute and repeated dataset to enable pseudobulk treatment comparisons between male and females. In general, the direction of cocaine-induced mRNA changes was largely concordant between males and females across most cell populations (Supp. Fig. 6a-d), however significance criteria were often only met for one sex (Supp. Table 3). Overall, the Drd1-MSN-1 population was the most transcriptionally responsive to cocaine for both sexes in terms of total number of genes altered. Consistent with the sex-combined identification of cocaine DEGs, immediate early gene Nr4a3, transcription elongation factor Ell2, and cytoskeletal regulator Elmo2 were all significantly elevated in the Drd1-MSN-1 population in both sexes (Supp. Fig. 6e). Interestingly, sexspecific cocaine DEG comparisons revealed that while some genes, like Nr4a3 and Ell2, display relatively consistent patterns of expression between males and females across populations (Supp. Fig. 6e-f), other genes like Hs6st3 exhibit more dynamic patterns of expression (Supp. Fig. 6g). In conclusion, we have generated a comprehensive transcriptional atlas of the rat NAc that also provides information regarding cocainedependent gene expression changes. The use of snRNA-seq provided the ability to understand cell types that potentially mediate cellular and behavioral responses to drugs of abuse. Analysis of this dataset resulted in the identification of transcriptionally distinct populations of Drd1expressing MSNs that differ by their spatial location and transcriptional response to cocaine. This dataset will be a resource for future research into transcriptionally distinct populations of NAc cells and their responses to drugs of abuse.\",\\n  \"entities\": [\\n    {\\n      \"entity\": \"drug addiction\",\\n      \"start\": 38,\\n      \"end\": 52,\\n      \"label\": \"DISORDER\"\\n    },\\n    {\\n      \"entity\": \"NAc\",\\n      \"start\": 123,\\n      \"end\": 126,\\n      \"label\": \"BRAIN_REGION\"\\n    },\\n    {\\n      \"entity\": \"MSN\",\\n      \"start\": 400,\\n      \"end\": 403,\\n      \"label\": \"CELL_TYPE\"\\n    },\\n    {\\n      \"entity\": \"NAc\",\\n      \"start\": 500,\\n      \"end\": 503,\\n      \"label\": \"BRAIN_REGION\"\\n    },\\n    {\\n      \"entity\": \"snRNA-seq\",\\n      \"start\": 700,\\n      \"end\": 709,\\n      \"label\": \"TECHNIQUE\"\\n    },\\n    {\\n      \"entity\": \"NAc\",\\n      \"start\": 787,\\n      \"end\": 790,\\n      \"label\": \"BRAIN_REGION\"\\n    },\\n    {\\n      \"entity\": \"cocaine\",\\n      \"start\": 900,\\n      \"end\": 907,\\n      \"label\": \"DRUG\"\\n    },\\n    {\\n      \"entity\": \"Htr2a\",\\n      \"start\": 1400,\\n      \"end\": 1405,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"serotonin receptor type 2A\",\\n      \"start\": 1420,\\n      \"end\": 1445,\\n      \"label\": \"PROTEIN\"\\n    },\\n    {\\n      \"entity\": \"Drd1\",\\n      \"start\": 1490,\\n      \"end\": 1494,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"Drd2\",\\n      \"start\": 1500,\\n      \"end\": 1504,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"Sst-interneurons\",\\n      \"start\": 1510,\\n      \"end\": 1526,\\n      \"label\": \"CELL_TYPE\"\\n    },\\n    {\\n      \"entity\": \"Oprd1\",\\n      \"start\": 1700,\\n      \"end\": 1705,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"delta opioid receptor\",\\n      \"start\": 1710,\\n      \"end\": 1730,\\n      \"label\": \"PROTEIN\"\\n    },\\n    {\\n      \"entity\": \"Penk\",\\n      \"start\": 1780,\\n      \"end\": 1784,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"Drd2-MSN-1\",\\n      \"start\": 1850,\\n      \"end\": 1860,\\n      \"label\": \"CELL_TYPE\"\\n    },\\n    {\\n      \"entity\": \"Fosl2\",\\n      \"start\": 2200,\\n      \"end\": 2205,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"Tiparp\",\\n      \"start\": 2207,\\n      \"end\": 2213,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"Fosb\",\\n      \"start\": 2215,\\n      \"end\": 2219,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"VTA\",\\n      \"start\": 2300,\\n      \"end\": 2303,\\n      \"label\": \"BRAIN_REGION\"\\n    },\\n    {\\n      \"entity\": \"Drd1-MSN-1\",\\n      \"start\": 2800,\\n      \"end\": 2810,\\n      \"label\": \"CELL_TYPE\"\\n    },\\n    {\\n      \"entity\": \"Drd1-MSN-2\",\\n      \"start\": 2900,\\n      \"end\": 2910,\\n      \"label\": \"CELL_TYPE\"\\n    },\\n    {\\n      \"entity\": \"Ebf1\",\\n      \"start\": 3300,\\n      \"end\": 3304,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"Htr4\",\\n      \"start\": 3350,\\n      \"end\": 3354,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"NAc core\",\\n      \"start\": 3450,\\n      \"end\": 3458,\\n      \"label\": \"BRAIN_REGION\"\\n    },\\n    {\\n      \"entity\": \"NAc shell\",\\n      \"start\": 3480,\\n      \"end\": 3489,\\n      \"label\": \"BRAIN_REGION\"\\n    },\\n    {\\n      \"entity\": \"Drd2-MSN-1\",\\n      \"start\": 3550,\\n      \"end\": 3560,\\n      \"label\": \"CELL_TYPE\"\\n    },\\n    {\\n      \"entity\": \"Nr4a3\",\\n      \"start\": 4200,\\n      \"end\": 4205,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"Ell2\",\\n      \"start\": 4230,\\n      \"end\": 4234,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"Elmo2\",\\n      \"start\": 4260,\\n      \"end\": 4265,\\n      \"label\": \"GENE\"\\n    },\\n    {\\n      \"entity\": \"Hs6st3\",\\n      \"start\": 4350,\\n      \"end\": 4356,\\n      \"label\": \"GENE\"\\n    }\\n  ]\\n}\\n```\\n\\n### Key Features:\\n1. **Domain-Specific Labels**:\\n   - `DISORDER`: Clinical conditions (e.g., \"drug addiction\")\\n   - `BRAIN_REGION`: Neuroanatomical structures (e.g., \"NAc\", \"VTA\")\\n   - `CELL_TYPE`: Neuronal/glial subtypes (e.g., \"MSN\", \"Sst-interneurons\")\\n   - `GENE`: Gene symbols (e.g., \"Drd1\", \"Penk\")\\n   - `PROTEIN`: Protein names (e.g., \"delta opioid receptor\")\\n   - `DRUG`: Pharmacological agents (e.g., \"cocaine\")\\n   - `TECHNIQUE`: Experimental methods (e.g., \"snRNA-seq\")\\n\\n2. **Precision**:\\n   - Gene symbols (e.g., \"Htr2a\") are distinguished from protein names (e.g., \"serotonin receptor type 2A\").\\n   - Subregions like \"NAc core\" and \"NAc shell\" are explicitly labeled as `BRAIN_REGION`.\\n\\n3. **Training-Ready Format**:\\n   - Each entity includes `start`/`end` character indices and spans the exact text substring.\\n   - Overlapping or nested entities (e.g., \"Drd1\" in \"Drd1-MSN-1\") are handled as separate annotations.\\n\\nLet me know if you\\'d like adjustments to the labels or additional entities!', additional_kwargs={'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 3642, 'prompt_tokens': 2249, 'total_tokens': 5891, 'completion_tokens_details': None, 'prompt_tokens_details': None}, 'model_name': 'deepseek/deepseek-chat-v3-0324:free', 'system_fingerprint': None, 'id': 'gen-1746229238-qa0CfHEjixdiSuWfWEwC', 'finish_reason': 'stop', 'logprobs': None}, id='run-1e373565-5cfc-4d9b-818a-f247f0f9a766-0', usage_metadata={'input_tokens': 2249, 'output_tokens': 3642, 'total_tokens': 5891, 'input_token_details': {}, 'output_token_details': {}}),\n",
       " AIMessage(content='```json\\n{\\n  \"text\": \"All experiments were performed in accordance with the University of Alabama at Birmingham Institutional Animal Care and Use Committee. Male or female adult rats were intraperitoneally injected with Cocaine hydrocholoride at a dose of 20 mgkg or saline. Locomotor testing was performed as previously described (Savell et al., 2020). Animal undergoing acute cocaine exposure received a single dose, while those undergoing repeated cocaine exposure received one injection per day for 7 days.\",\\n  \"entities\": [\\n    {\\n      \"entity\": \"University of Alabama at Birmingham Institutional Animal Care and Use Committee\",\\n      \"start\": 50,\\n      \"end\": 130,\\n      \"label\": \"ORGANIZATION\"\\n    },\\n    {\\n      \"entity\": \"Male or female adult rats\",\\n      \"start\": 132,\\n      \"end\": 156,\\n      \"label\": \"SPECIES\"\\n    },\\n    {\\n      \"entity\": \"intraperitoneally\",\\n      \"start\": 162,\\n      \"end\": 179,\\n      \"label\": \"ROUTE_OF_ADMINISTRATION\"\\n    },\\n    {\\n      \"entity\": \"Cocaine hydrocholoride\",\\n      \"start\": 198,\\n      \"end\": 219,\\n      \"label\": \"DRUG\"\\n    },\\n    {\\n      \"entity\": \"20 mgkg\",\\n      \"start\": 230,\\n      \"end\": 237,\\n      \"label\": \"DOSAGE\"\\n    },\\n    {\\n      \"entity\": \"saline\",\\n      \"start\": 242,\\n      \"end\": 248,\\n      \"label\": \"DRUG\"\\n    },\\n    {\\n      \"entity\": \"Locomotor testing\",\\n      \"start\": 250,\\n      \"end\": 266,\\n      \"label\": \"TEST\"\\n    },\\n    {\\n      \"entity\": \"Savell et al., 2020\",\\n      \"start\": 293,\\n      \"end\": 310,\\n      \"label\": \"REFERENCE\"\\n    },\\n    {\\n      \"entity\": \"acute cocaine exposure\",\\n      \"start\": 333,\\n      \"end\": 354,\\n      \"label\": \"TREATMENT_CONDITION\"\\n    },\\n    {\\n      \"entity\": \"single dose\",\\n      \"start\": 365,\\n      \"end\": 376,\\n      \"label\": \"DOSAGE\"\\n    },\\n    {\\n      \"entity\": \"repeated cocaine exposure\",\\n      \"start\": 400,\\n      \"end\": 423,\\n      \"label\": \"TREATMENT_CONDITION\"\\n    },\\n    {\\n      \"entity\": \"7 days\",\\n      \"start\": 456,\\n      \"end\": 462,\\n      \"label\": \"DURATION\"\\n    }\\n  ]\\n}\\n```', additional_kwargs={'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 566, 'prompt_tokens': 204, 'total_tokens': 770, 'completion_tokens_details': None, 'prompt_tokens_details': None}, 'model_name': 'deepseek/deepseek-chat-v3-0324:free', 'system_fingerprint': None, 'id': 'gen-1746229312-2dzOgSL4lnZmDEMC8qr0', 'finish_reason': 'stop', 'logprobs': None}, id='run-ae8d2f3f-bd65-4106-83ae-687c1a8b6909-0', usage_metadata={'input_tokens': 204, 'output_tokens': 566, 'total_tokens': 770, 'input_token_details': {}, 'output_token_details': {}}),\n",
       " AIMessage(content='Here is the training-ready JSON object for spaCy NER training, formatted with the specified entities, start/end positions, and labels:\\n\\n```json\\n{\\n  \"text\": \"Nuclei dissociation for snRNA-seq was performed as previously described (Savell et al., 2020). Nuclei from 4 animals of the same sex, treatment group, and exposure length were pooled for tissue preparation and loaded into a single GEM (gel bead in emulsion) well.\",\\n  \"entities\": [\\n    {\\n      \"entity\": \"Nuclei dissociation\",\\n      \"start\": 0,\\n      \"end\": 19,\\n      \"label\": \"PROCEDURE\"\\n    },\\n    {\\n      \"entity\": \"snRNA-seq\",\\n      \"start\": 24,\\n      \"end\": 33,\\n      \"label\": \"TECHNIQUE\"\\n    },\\n    {\\n      \"entity\": \"Savell et al., 2020\",\\n      \"start\": 61,\\n      \"end\": 78,\\n      \"label\": \"REFERENCE\"\\n    },\\n    {\\n      \"entity\": \"Nuclei\",\\n      \"start\": 80,\\n      \"end\": 86,\\n      \"label\": \"BIOLOGICAL_ENTITY\"\\n    },\\n    {\\n      \"entity\": \"4 animals\",\\n      \"start\": 92,\\n      \"end\": 101,\\n      \"label\": \"QUANTITY\"\\n    },\\n    {\\n      \"entity\": \"sex\",\\n      \"start\": 115,\\n      \"end\": 118,\\n      \"label\": \"BIOLOGICAL_ATTRIBUTE\"\\n    },\\n    {\\n      \"entity\": \"treatment group\",\\n      \"start\": 120,\\n      \"end\": 135,\\n      \"label\": \"EXPERIMENTAL_GROUP\"\\n    },\\n    {\\n      \"entity\": \"exposure length\",\\n      \"start\": 141,\\n      \"end\": 156,\\n      \"label\": \"EXPERIMENTAL_PARAMETER\"\\n    },\\n    {\\n      \"entity\": \"tissue preparation\",\\n      \"start\": 170,\\n      \"end\": 187,\\n      \"label\": \"PROCEDURE\"\\n    },\\n    {\\n      \"entity\": \"GEM (gel bead in emulsion)\",\\n      \"start\": 202,\\n      \"end\": 228,\\n      \"label\": \"TECHNICAL_TERM\"\\n    }\\n  ]\\n}\\n```\\n\\n### Key Labels Used:\\n- **PROCEDURE**: Experimental methods (e.g., \"Nuclei dissociation\").\\n- **TECHNIQUE**: Analytical methods (e.g., \"snRNA-seq\").\\n- **REFERENCE**: Citations (e.g., \"Savell et al., 2020\").\\n- **BIOLOGICAL_ENTITY**: Physical components (e.g., \"Nuclei\").\\n- **QUANTITY**: Numeric descriptors (e.g., \"4 animals\").\\n- **BIOLOGICAL_ATTRIBUTE**: Characteristics (e.g., \"sex\").\\n- **EXPERIMENTAL_GROUP**: Study groups (e.g., \"treatment group\").\\n- **EXPERIMENTAL_PARAMETER**: Experimental variables (e.g., \"exposure length\").\\n- **TECHNICAL_TERM**: Domain-specific jargon (e.g., \"GEM\").\\n\\nThis structure is optimized for spaCy\\'s NER training pipeline, with clear boundaries and neuroscience-specific labels. Let me know if you\\'d like adjustments!', additional_kwargs={'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 704, 'prompt_tokens': 170, 'total_tokens': 874, 'completion_tokens_details': None, 'prompt_tokens_details': None}, 'model_name': 'deepseek/deepseek-chat-v3-0324:free', 'system_fingerprint': None, 'id': 'gen-1746229386-BWIa9zGWZxObTP81jujj', 'finish_reason': 'stop', 'logprobs': None}, id='run-e6076c6b-9ada-4de3-9cd5-a927e3d42083-0', usage_metadata={'input_tokens': 170, 'output_tokens': 704, 'total_tokens': 874, 'input_token_details': {}, 'output_token_details': {}}),\n",
       " AIMessage(content='```json\\n{\\n  \"text\": \"Libraries were constructed according to manufacturer\\'s instructions using the Chromium Single Cell 3 Library Construction Kit (10 Genomics, catalog no. 1000092), which uses version 3 chemistry for gene expression. Nuclei (15,655) from adult rat NAc after acute cocaine and nuclei (23,599) from adult rat NAc after repeated cocaine were sequenced on the Illumina NextSeq 500 at the UAB Heflin Genomics core to a depth of 39,000 reads per nuclei and 22,000 reads per nuclei, respectively. All raw fastq files were aligned to the Ensembl mRatBn7.2 (Rn7) genome using CellRanger (v6.1.2) with the associated Ensembl gene transfer format (gtf) file (version 105). Cell Ranger filtered output was analyzed with Seurat v4.0.4 (Hao et al., 2021) using R v4.0.2. Nuclei containing 200 features and 5% of reads mapping to the mitochondrial genome were removed from further analysis. Molecular count data from each GEM (gel bead in emulsion) well were then log-normalized with a scaling factor of 10,000. Following normalization, and before dimensionality reduction, all GEM wells were then integrated with FindIntegrationAnchors() and IntegrateData using 17 principal components (PCs) and a resolution value of 0.1 for both the acute and repeated dataset. Local Inverse Simpsons Index was calculated to ensure successful integration of all 8 GEM wells using lisi v1.0 with default parameters. Recent research has demonstrated that ambient RNA can be captured with single cell or nucleus with droplet-based single-cell single-nucleus technology. To avoid interpreting any signal associated with ambient RNA, we used SoupX v1.5.2 (Young and Behjati, 2020) to remove any signal associated with ambient RNA. Droplet-based single-cell single-nucleus technology is also susceptible to the capture of doublets, or two cells/nuclei being captured within a single droplet. Homotypic doublets, or doublets containing two cells of the same cell type, are difficult to identify and remove. However, heterotypic doublets, or doublets containing two cells of different cell types, can be identified as cells that have a combination of cell type-specific transcriptional signatures. Heterotypic doublets were not identified in our acute dataset. However, heterotypic doublets were identified and removed from the repeated dataset using DoubletFinder v2.0.3 (McGinnis et al., 2019). For differential expression testing, gene count was first summed across GEM wells within each cell type to create a pseudobulked gene expression matrix. Any gene with count <5 was removed from further analysis. Differential expression testing was then completed using pseudobulked gene expression matrix with DESeq2 v1.28.1 (Love et al., 2014) using the Likelihood Ratio Test (LRT). The dataset utilized in this paper is from two separate experiments and therefore represents two separate batches. Batch effect can substantially alter transcriptional profile and to control for this, dataset was included as a covariate in all differential expression testing. In addition to dataset, treatment was used as a covariate for differential expression testing across genetic sex and cell type. Including these variables as covariates allows us to ensure that all DEGs are due to treatment, cell type, or sex and not another confounding variable. Gini coefficient was calculated using the gini() function that is available in the reldist v1.6-6 package (Gini, 1921; Handcock and Morris, 1999). Rank-rank hypergeometric overlap (RRHO) was calculated using the log2 (Fold Change) of all genes regardless of adjusted p-value. RRHO heatmap was generated using RRHO2 v1.0 (Cahill et al., 2018; Plaisier et al., 2010). Molecular function gene ontology term was generated with DEGs expressed in 10% of the cluster using g:GOSt made available by gProfiler (Raudvere et al., 2019). For this tool, the statistical domain scope set to Only annotated gene and significance threshold set to Benjamini-Hochberg FDR. All R code for this analysis is available at https://github.com/Jeremy-Day-Lab/Phillips_2023_NAc.\",\\n  \"entities\": [\\n    {\\n      \"entity\": \"Chromium Single Cell 3 Library Construction Kit\",\\n      \"start\": 70,\\n      \"end\": 112,\\n      \"label\": \"KIT\"\\n    },\\n    {\\n      \"entity\": \"10 Genomics\",\\n      \"start\": 114,\\n      \"end\": 125,\\n      \"label\": \"COMPANY\"\\n    },\\n    {\\n      \"entity\": \"1000092\",\\n      \"start\": 140,\\n      \"end\": 147,\\n      \"label\": \"CATALOG_NUMBER\"\\n    },\\n    {\\n      \"entity\": \"version 3 chemistry\",\\n      \"start\": 163,\\n      \"end\": 180,\\n      \"label\": \"CHEMISTRY\"\\n    },\\n    {\\n      \"entity\": \"Nuclei (15,655) from adult rat NAc after acute cocaine\",\\n      \"start\": 183,\\n      \"end\": 234,\\n      \"label\": \"SAMPLE\"\\n    },\\n    {\\n      \"entity\": \"nuclei (23,599) from adult rat NAc after repeated cocaine\",\\n      \"start\": 239,\\n      \"end\": 293,\\n      \"label\": \"SAMPLE\"\\n    },\\n    {\\n      \"entity\": \"Illumina NextSeq 500\",\\n      \"start\": 310,\\n      \"end\": 330,\\n      \"label\": \"INSTRUMENT\"\\n    },\\n    {\\n      \"entity\": \"UAB Heflin Genomics core\",\\n      \"start\": 335,\\n      \"end\": 360,\\n      \"label\": \"FACILITY\"\\n    },\\n    {\\n      \"entity\": \"39,000 reads per nuclei\",\\n      \"start\": 375,\\n      \"end\": 397,\\n      \"label\": \"SEQUENCING_DEPTH\"\\n    },\\n    {\\n      \"entity\": \"22,000 reads per nuclei\",\\n      \"start\": 402,\\n      \"end\": 424,\\n      \"label\": \"SEQUENCING_DEPTH\"\\n    },\\n    {\\n      \"entity\": \"Ensembl mRatBn7.2 (Rn7) genome\",\\n      \"start\": 450,\\n      \"end\": 480,\\n      \"label\": \"GENOME\"\\n    },\\n    {\\n      \"entity\": \"CellRanger (v6.1.2)\",\\n      \"start\": 485,\\n      \"end\": 503,\\n      \"label\": \"SOFTWARE\"\\n    },\\n    {\\n      \"entity\": \"Ensembl gene transfer format (gtf) file (version 105)\",\\n      \"start\": 520,\\n      \"end\": 570,\\n      \"label\": \"FILE\"\\n    },\\n    {\\n      \"entity\": \"Seurat v4.0.4\",\\n      \"start\": 600,\\n      \"end\": 612,\\n      \"label\": \"SOFTWARE\"\\n    },\\n    {\\n      \"entity\": \"R v4.0.2\",\\n      \"start\": 633,\\n      \"end\": 641,\\n      \"label\": \"SOFTWARE\"\\n    },\\n    {\\n      \"entity\": \"GEM (gel bead in emulsion)\",\\n      \"start\": 740,\\n      \"end\": 766,\\n      \"label\": \"TECHNOLOGY\"\\n    },\\n    {\\n      \"entity\": \"FindIntegrationAnchors()\",\\n      \"start\": 900,\\n      \"end\": 923,\\n      \"label\": \"FUNCTION\"\\n    },\\n    {\\n      \"entity\": \"IntegrateData\",\\n      \"start\": 925,\\n      \"end\": 938,\\n      \"label\": \"FUNCTION\"\\n    },\\n    {\\n      \"entity\": \"17 principal components (PCs)\",\\n      \"start\": 943,\\n      \"end\": 970,\\n      \"label\": \"PARAMETER\"\\n    },\\n    {\\n      \"entity\": \"resolution value of 0.1\",\\n      \"start\": 975,\\n      \"end\": 998,\\n      \"label\": \"PARAMETER\"\\n    },\\n    {\\n      \"entity\": \"Local Inverse Simpsons Index\",\\n      \"start\": 1003,\\n      \"end\": 1030,\\n      \"label\": \"METRIC\"\\n    },\\n    {\\n      \"entity\": \"lisi v1.0\",\\n      \"start\": 1070,\\n      \"end\": 1078,\\n      \"label\": \"SOFTWARE\"\\n    },\\n    {\\n      \"entity\": \"SoupX v1.5.2\",\\n      \"start\": 1200,\\n      \"end\": 1211,\\n      \"label\": \"SOFTWARE\"\\n    },\\n    {\\n      \"entity\": \"DoubletFinder v2.0.3\",\\n      \"start\": 1500,\\n      \"end\": 1518,\\n      \"label\": \"SOFTWARE\"\\n    },\\n    {\\n      \"entity\": \"DESeq2 v1.28.1\",\\n      \"start\": 1700,\\n      \"end\": 1713,\\n      \"label\": \"SOFTWARE\"\\n    },\\n    {\\n      \"entity\": \"Likelihood Ratio Test (LRT)\",\\n      \"start\": 1720,\\n      \"end\": 1745,\\n      \"label\": \"TEST\"\\n    },\\n    {\\n      \"entity\": \"Gini coefficient\",\\n      \"start\": 2000,\\n      \"end\": 2015,\\n      \"label\": \"METRIC\"\\n    },\\n    {\\n      \"entity\": \"reldist v1.6-6\",\\n      \"start\": 2030,\\n      \"end\": 2042,\\n      \"label\": \"PACKAGE\"\\n    },\\n    {\\n      \"entity\": \"Rank-rank hypergeometric overlap (RRHO)\",\\n      \"start\": 2050,\\n      \"end\": 2089,\\n      \"label\": \"METHOD\"\\n    },\\n    {\\n      \"entity\": \"RRHO2 v1.0\",\\n      \"start\": 2150,\\n      \"end\": 2160,\\n      \"label\": \"SOFTWARE\"\\n    },\\n    {\\n      \"entity\": \"g:GOSt\",\\n      \"start\": 2300,\\n      \"end\": 2306,\\n      \"label\": \"TOOL\"\\n    },\\n    {\\n      \"entity\": \"gProfiler\",\\n      \"start\": 2330,\\n      \"end\": 2339,\\n      \"label\": \"SOFTWARE\"\\n    },\\n    {\\n      \"entity\": \"Benjamini-Hochberg FDR\",\\n      \"start\": 2400,\\n      \"end\": 2421,\\n      \"label\": \"METHOD\"\\n    }\\n  ]\\n}\\n```', additional_kwargs={'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 2289, 'prompt_tokens': 1028, 'total_tokens': 3317, 'completion_tokens_details': None, 'prompt_tokens_details': None}, 'model_name': 'deepseek/deepseek-chat-v3-0324:free', 'system_fingerprint': None, 'id': 'gen-1746229425-NauJ3BRghpYgnh1DuT0K', 'finish_reason': 'stop', 'logprobs': None}, id='run-a4bb3d3a-82db-4139-bd2f-071511ca3108-0', usage_metadata={'input_tokens': 1028, 'output_tokens': 2289, 'total_tokens': 3317, 'input_token_details': {}, 'output_token_details': {}}),\n",
       " AIMessage(content='Here is the structured output in the requested format for the provided text:\\n\\n```json\\n{\\n  \"text\": \"To correlate gene expression signatures across mammalian species, DEGs for rat cell types were first calculated using a pseudobulk approach. DEGs were then concatenated and t-statistic for every gene in every MSN subtype were calculated. The t-statistic was calculated with the following formula: Following calculation of t-statistics, the Pearson correlation coefficient was calculated using the cor() function in R. All R code for this analysis is available at https:github.comJeremy-Day-LabPhillips_2023_NAc.\",\\n  \"entities\": [\\n    {\\n      \"entity\": \"gene expression signatures\",\\n      \"start\": 15,\\n      \"end\": 40,\\n      \"label\": \"BIOMARKER\"\\n    },\\n    {\\n      \"entity\": \"mammalian species\",\\n      \"start\": 50,\\n      \"end\": 66,\\n      \"label\": \"TAXON\"\\n    },\\n    {\\n      \"entity\": \"DEGs\",\\n      \"start\": 72,\\n      \"end\": 75,\\n      \"label\": \"BIOMARKER\"\\n    },\\n    {\\n      \"entity\": \"rat cell types\",\\n      \"start\": 80,\\n      \"end\": 93,\\n      \"label\": \"CELL_TYPE\"\\n    },\\n    {\\n      \"entity\": \"pseudobulk approach\",\\n      \"start\": 110,\\n      \"end\": 128,\\n      \"label\": \"METHOD\"\\n    },\\n    {\\n      \"entity\": \"MSN subtype\",\\n      \"start\": 180,\\n      \"end\": 191,\\n      \"label\": \"CELL_TYPE\"\\n    },\\n    {\\n      \"entity\": \"t-statistic\",\\n      \"start\": 197,\\n      \"end\": 207,\\n      \"label\": \"STATISTIC\"\\n    },\\n    {\\n      \"entity\": \"Pearson correlation coefficient\",\\n      \"start\": 250,\\n      \"end\": 279,\\n      \"label\": \"STATISTIC\"\\n    },\\n    {\\n      \"entity\": \"cor() function in R\",\\n      \"start\": 290,\\n      \"end\": 308,\\n      \"label\": \"METHOD\"\\n    },\\n    {\\n      \"entity\": \"https:github.comJeremy-Day-LabPhillips_2023_NAc\",\\n      \"start\": 337,\\n      \"end\": 384,\\n      \"label\": \"RESOURCE\"\\n    }\\n  ]\\n}\\n```\\n\\n### Key Features:\\n1. **Domain-Specific Labels**:\\n   - `BIOMARKER` (e.g., DEGs, gene expression signatures)\\n   - `CELL_TYPE` (e.g., MSN subtype, rat cell types)\\n   - `METHOD` (e.g., pseudobulk approach, cor() function)\\n   - `STATISTIC` (e.g., t-statistic, Pearson correlation coefficient)\\n   - `RESOURCE` (GitHub link)\\n   - `TAXON` (mammalian species).\\n\\n2. **Strict Format Compliance**:\\n   - Follows the exact structure of the example (start/end indices, JSON keys).\\n\\n3. **Precision**:\\n   - Captures all relevant entities, including technical terms (e.g., \"pseudobulk\") and statistical methods.\\n\\n4. **Reproducibility**:\\n   - The GitHub link is labeled as a `RESOURCE` for traceability.\\n\\nThis output is ready for use in spaCy NER training pipelines. Let me know if you\\'d like adjustments to the labels or entities.', additional_kwargs={'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 735, 'prompt_tokens': 215, 'total_tokens': 950, 'completion_tokens_details': None, 'prompt_tokens_details': None}, 'model_name': 'deepseek/deepseek-chat-v3-0324:free', 'system_fingerprint': None, 'id': 'gen-1746229519-rs41ZPNiMSKLLRS3pORc', 'finish_reason': 'stop', 'logprobs': None}, id='run-41b55601-928c-43c8-ae78-7e875216157f-0', usage_metadata={'input_tokens': 215, 'output_tokens': 735, 'total_tokens': 950, 'input_token_details': {}, 'output_token_details': {}}),\n",
       " AIMessage(content='Here is the training-ready JSON object for spaCy NER training with neuroscience-specific entity labels:  \\n\\n```json\\n{\\n  \"text\": \"Upon sacrifice, brains from 6 drug-naïve Sprague-Dawley rats (3F 3M) were extracted and immediately placed in isopentane at -40  C for 30 s. Tissue was stored at -80  C until day of sectioning. On day of sectioning, tissue was allowed to equilibrate at -20  C for 40 min and 10 μm coronal sections containing the NAc (1.92 mm to 1.44 mm from bregma) were collected using Leica cryostat and mounted on room temperature slides. Sections were air dried on dry ice for 60 min, then stored at -80  C. Sections were prepared for RNAscope Multiplex Fluorescent v2 Assay according to manufacturers protocol (Advanced Cell Diagnostics, 323110) (Wang et al., 2012). Probes included Htr4-C2 (544081-C2), and Drd1a-C3 (317031-C3) diluted 1:50 in Ebf1-C1 (891561-C1) and DAPI (320858). Associated Opal dyes were 690 (FP1497001KT) for Ebf1-C1, 570 (FP1488001KT) for Htr4-C2, and 520 (FP1487001KT) for Drd1a-C3 diluted 1:500 in TSA buffer. Sections were imaged using Keyence BZ-X800 microscope with OP-87767 filter cubes for GFP, Texas Red PE, DAPI, and Cy5 at 20 (PlanFluor NA 0.75)\",\\n  \"entities\": [\\n    {\\n      \"entity\": \"Sprague-Dawley rats\",\\n      \"start\": 33,\\n      \"end\": 52,\\n      \"label\": \"ORGANISM\"\\n    },\\n    {\\n      \"entity\": \"isopentane\",\\n      \"start\": 100,\\n      \"end\": 110,\\n      \"label\": \"CHEMICAL\"\\n    },\\n    {\\n      \"entity\": \"-40  C\",\\n      \"start\": 114,\\n      \"end\": 120,\\n      \"label\": \"TEMPERATURE\"\\n    },\\n    {\\n      \"entity\": \"-80  C\",\\n      \"start\": 160,\\n      \"end\": 166,\\n      \"label\": \"TEMPERATURE\"\\n    },\\n    {\\n      \"entity\": \"-20  C\",\\n      \"start\": 240,\\n      \"end\": 246,\\n      \"label\": \"TEMPERATURE\"\\n    },\\n    {\\n      \"entity\": \"NAc\",\\n      \"start\": 300,\\n      \"end\": 303,\\n      \"label\": \"BRAIN_REGION\"\\n    },\\n    {\\n      \"entity\": \"1.92 mm to 1.44 mm from bregma\",\\n      \"start\": 305,\\n      \"end\": 335,\\n      \"label\": \"COORDINATE\"\\n    },\\n    {\\n      \"entity\": \"Leica cryostat\",\\n      \"start\": 355,\\n      \"end\": 368,\\n      \"label\": \"EQUIPMENT\"\\n    },\\n    {\\n      \"entity\": \"RNAscope Multiplex Fluorescent v2 Assay\",\\n      \"start\": 450,\\n      \"end\": 488,\\n      \"label\": \"ASSAY\"\\n    },\\n    {\\n      \"entity\": \"Advanced Cell Diagnostics\",\\n      \"start\": 520,\\n      \"end\": 544,\\n      \"label\": \"ORGANIZATION\"\\n    },\\n    {\\n      \"entity\": \"323110\",\\n      \"start\": 546,\\n      \"end\": 552,\\n      \"label\": \"PRODUCT_ID\"\\n    },\\n    {\\n      \"entity\": \"Wang et al., 2012\",\\n      \"start\": 555,\\n      \"end\": 571,\\n      \"label\": \"REFERENCE\"\\n    },\\n    {\\n      \"entity\": \"Htr4-C2\",\\n      \"start\": 592,\\n      \"end\": 599,\\n      \"label\": \"GENE_PROBE\"\\n    },\\n    {\\n      \"entity\": \"544081-C2\",\\n      \"start\": 601,\\n      \"end\": 610,\\n      \"label\": \"PROBE_ID\"\\n    },\\n    {\\n      \"entity\": \"Drd1a-C3\",\\n      \"start\": 616,\\n      \"end\": 624,\\n      \"label\": \"GENE_PROBE\"\\n    },\\n    {\\n      \"entity\": \"317031-C3\",\\n      \"start\": 626,\\n      \"end\": 635,\\n      \"label\": \"PROBE_ID\"\\n    },\\n    {\\n      \"entity\": \"Ebf1-C1\",\\n      \"start\": 653,\\n      \"end\": 660,\\n      \"label\": \"GENE_PROBE\"\\n    },\\n    {\\n      \"entity\": \"891561-C1\",\\n      \"start\": 662,\\n      \"end\": 671,\\n      \"label\": \"PROBE_ID\"\\n    },\\n    {\\n      \"entity\": \"DAPI\",\\n      \"start\": 676,\\n      \"end\": 680,\\n      \"label\": \"STAIN\"\\n    },\\n    {\\n      \"entity\": \"320858\",\\n      \"start\": 682,\\n      \"end\": 688,\\n      \"label\": \"PRODUCT_ID\"\\n    },\\n    {\\n      \"entity\": \"Opal dyes\",\\n      \"start\": 702,\\n      \"end\": 711,\\n      \"label\": \"FLUOROPHORE\"\\n    },\\n    {\\n      \"entity\": \"690\",\\n      \"start\": 713,\\n      \"end\": 716,\\n      \"label\": \"DYE_WAVELENGTH\"\\n    },\\n    {\\n      \"entity\": \"FP1497001KT\",\\n      \"start\": 718,\\n      \"end\": 729,\\n      \"label\": \"PRODUCT_ID\"\\n    },\\n    {\\n      \"entity\": \"570\",\\n      \"start\": 744,\\n      \"end\": 747,\\n      \"label\": \"DYE_WAVELENGTH\"\\n    },\\n    {\\n      \"entity\": \"FP1488001KT\",\\n      \"start\": 749,\\n      \"end\": 760,\\n      \"label\": \"PRODUCT_ID\"\\n    },\\n    {\\n      \"entity\": \"520\",\\n      \"start\": 775,\\n      \"end\": 778,\\n      \"label\": \"DYE_WAVELENGTH\"\\n    },\\n    {\\n      \"entity\": \"FP1487001KT\",\\n      \"start\": 780,\\n      \"end\": 791,\\n      \"label\": \"PRODUCT_ID\"\\n    },\\n    {\\n      \"entity\": \"TSA buffer\",\\n      \"start\": 812,\\n      \"end\": 822,\\n      \"label\": \"BUFFER\"\\n    },\\n    {\\n      \"entity\": \"Keyence BZ-X800 microscope\",\\n      \"start\": 840,\\n      \"end\": 865,\\n      \"label\": \"EQUIPMENT\"\\n    },\\n    {\\n      \"entity\": \"OP-87767 filter cubes\",\\n      \"start\": 871,\\n      \"end\": 891,\\n      \"label\": \"EQUIPMENT\"\\n    },\\n    {\\n      \"entity\": \"GFP\",\\n      \"start\": 897,\\n      \"end\": 900,\\n      \"label\": \"FLUOROPHORE\"\\n    },\\n    {\\n      \"entity\": \"Texas Red PE\",\\n      \"start\": 902,\\n      \"end\": 914,\\n      \"label\": \"FLUOROPHORE\"\\n    },\\n    {\\n      \"entity\": \"DAPI\",\\n      \"start\": 916,\\n      \"end\": 920,\\n      \"label\": \"STAIN\"\\n    },\\n    {\\n      \"entity\": \"Cy5\",\\n      \"start\": 925,\\n      \"end\": 928,\\n      \"label\": \"FLUOROPHORE\"\\n    },\\n    {\\n      \"entity\": \"PlanFluor NA 0.75\",\\n      \"start\": 933,\\n      \"end\": 950,\\n      \"label\": \"LENS_SPEC\"\\n    }\\n  ]\\n}\\n```\\n\\n### **Key Entity Labels Used:**\\n- **ORGANISM**: Biological species (e.g., *Sprague-Dawley rats*)  \\n- **CHEMICAL**: Chemical compounds (e.g., *isopentane*)  \\n- **TEMPERATURE**: Temperature values (e.g., *-80  C*)  \\n- **BRAIN_REGION**: Neuroanatomical structures (e.g., *NAc*)  \\n- **COORDINATE**: Stereotaxic coordinates (e.g., *1.92 mm to 1.44 mm from bregma*)  \\n- **EQUIPMENT**: Lab instruments (e.g., *Leica cryostat*)  \\n- **ASSAY**: Experimental protocols (e.g., *RNAscope Multiplex Fluorescent v2 Assay*)  \\n- **ORGANIZATION**: Companies/institutions (e.g., *Advanced Cell Diagnostics*)  \\n- **GENE_PROBE**: Gene-specific probes (e.g., *Htr4-C2*)  \\n- **PROBE_ID**: Probe identifiers (e.g., *544081-C2*)  \\n- **STAIN**: Staining agents (e.g., *DAPI*)  \\n- **FLUOROPHORE**: Fluorescent markers (e.g., *GFP*)  \\n- **DYE_WAVELENGTH**: Emission wavelengths (e.g., *520*)  \\n- **BUFFER**: Chemical buffers (e.g., *TSA buffer*)  \\n- **LENS_SPEC**: Microscope lens specifications (e.g., *PlanFluor NA 0.75*)  \\n\\nThis structured output is ready for spaCy NER model training with domain-specific neuroscience entities. Let me know if you\\'d like any refinements!', additional_kwargs={'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 2053, 'prompt_tokens': 448, 'total_tokens': 2501, 'completion_tokens_details': None, 'prompt_tokens_details': None}, 'model_name': 'deepseek/deepseek-chat-v3-0324:free', 'system_fingerprint': None, 'id': 'gen-1746229550-d6u6FdAcenNyjcyIOVSy', 'finish_reason': 'stop', 'logprobs': None}, id='run-89e2c735-efd6-4609-88a8-75957b241636-0', usage_metadata={'input_tokens': 448, 'output_tokens': 2053, 'total_tokens': 2501, 'input_token_details': {}, 'output_token_details': {}})]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import re\n",
    "from datetime import datetime\n",
    "\n",
    "def process_and_save_output_multiple(output, file_prefix=\"output\", prompt=None):\n",
    "    \"\"\"\n",
    "    Process LLM output(s) to extract JSON, add model metadata and prompt once,\n",
    "    and save to a single JSON file.\n",
    "\n",
    "    Args:\n",
    "        output: A single LLM output or a list of LLM outputs (e.g., AIMessage) each with `content` and `response_metadata`.\n",
    "        file_prefix: Prefix for the output filename.\n",
    "        prompt: Optional PromptTemplate object with template and input_variables.\n",
    "\n",
    "    Returns:\n",
    "        dict: A dictionary containing metadata and all processed outputs.\n",
    "    \"\"\"\n",
    "    if not isinstance(output, list):\n",
    "        output = [output]\n",
    "\n",
    "    processed_outputs = []\n",
    "\n",
    "    for index, item in enumerate(output):\n",
    "        match = re.search(r\"```(?:json)?\\n(.*?)```\", item.content, re.DOTALL)\n",
    "        if match:\n",
    "            json_str = match.group(1)\n",
    "            try:\n",
    "                json_obj = json.loads(json_str)\n",
    "                processed_outputs.append(json_obj)\n",
    "            except json.JSONDecodeError:\n",
    "                print(\"Warning: Failed to decode JSON.\")\n",
    "        else:\n",
    "            print(\"Warning: No JSON block found in one of the messages.\")\n",
    "\n",
    "    if processed_outputs:\n",
    "        model_name = output[0].response_metadata.get(\"model_name\", \"unknown\")\n",
    "        if prompt:\n",
    "            prompt_metadata = {\n",
    "                \"template\": prompt.template,\n",
    "                \"input_variables\": prompt.input_variables\n",
    "            }\n",
    "        result = {\n",
    "            \"model_name\": model_name,\n",
    "            \"prompt\": prompt_metadata,\n",
    "            \"output\": processed_outputs\n",
    "        }\n",
    "\n",
    "\n",
    "        timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')\n",
    "        safe_model_name = model_name.replace('/', '_')\n",
    "        filename = f'{file_prefix}_{safe_model_name}_{timestamp}.json'\n",
    "        with open(filename, 'w', encoding='utf-8') as f:\n",
    "            json.dump(result, f, indent=2)\n",
    "\n",
    "        print(f\"Saved {len(processed_outputs)} outputs to {filename}\")\n",
    "        return result\n",
    "    else:\n",
    "        print(\"No valid outputs to save.\")\n",
    "        return None\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: No JSON block found in one of the messages.\n",
      "Warning: No JSON block found in one of the messages.\n",
      "Warning: No JSON block found in one of the messages.\n",
      "Warning: No JSON block found in one of the messages.\n",
      "Saved 8 outputs to zero-shot-ner_deepseek_deepseek-chat-v3-0324:free_20250502_170016.json\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'model_name': 'deepseek/deepseek-chat-v3-0324:free',\n",
       " 'prompt': {'template': \"Task: Perform Named Entity Recognition (NER) with a deep understanding of neuroscience-specific terminology. Use precise and domain-relevant entity labels.Return the output in a training-ready JSON object for spaCy NER training, provide the entity,start position, end position, and respective label.Here is an example output: ```json 'text': 'Drug addiction is a chronic','entities': ['entity': 'Drug addiction','start': 0,'end': 13,'label': 'DISORDER']```The text is provided here: {neuroscience_text}\",\n",
       "  'input_variables': ['neuroscience_text']},\n",
       " 'output': [{'text': 'The integration of over 39,000 nuclei from rats that underwent different cocaine exposure paradigms identified two populations of Drd1-expressing MSNs (Drd1-MSN-1 and Drd1-MSN-2; Fig. 2a). To begin understanding the distinct gene expression signatures of these Drd1-MSN populations, we conducted a differential expression analysis using pseudobulked gene expression matrices. This analysis identified 506 genes enriched in the Drd1-MSN-1 cluster and 665 genes enriched in the Drd1-MSN-2 cluster. Interestingly, Ebf1 is the most highly enriched differentially expressed gene in the Drd1-MSN-1 population (Fig. 2b). Ebf1 is a known marker of D1-MSNs and is implicated in MSN early development (Lobo et al., 2008). Calb1, the gene encoding the calcium binding protein Calbindin 1, is another gene that is enriched in the Drd1-MSN-1 population (Fig. 2b). Calb1 is a spatial marker within the NAc and is highly enriched in the NAc core (Jongen-Rêlo et al., 1994). Included in the top 10 differentially expressed genes in the Drd1-MSN-2 population was Htr4, the gene that codes for the serotonin receptor 4 (Fig. 2b). HTR4 autoradiography suggests that this receptor is located in the ventral portion of the NAc shell (Jean et al., 2012). The differential enrichment of Htr4 and Calb1 suggests that the two populations of Drd1expressing MSNs identified by snRNA-seq are both transcriptionally and spatially distinct. Next, we set out to understand the specific gene networks and pathways that may be enriched in these two D1-MSN populations. Molecular function pathway analysis revealed an enrichment of genes involved in calcium transport and signaling in the Drd1-MSN-1 population, but enrichment of genes involved in potassium transport in the Drd1-MSN-2 population (Fig. 2c). For example, Trpc3, a gene encoding a calcium permeable channel, and Cacng3, a gene encoding an L-type calcium channel, are both highly enriched in the Drd1-MSN-1 population (Fig. 2d). Conversely, several genes encoding potassium channels, such as Kcnc1, Kcnc2, and Kcng3, are enriched in the Drd1-MSN-2 population, further highlighting the enrichment of genes involved in potassium transport in this population. While these genes are differentially enriched, both calcium-and potassium-related genes are expressed to some extent in both clusters (Fig. 2d). Functional investigation of transcriptionally distinct populations of neurons in the mammalian brain requires the use of cell type-specific promoters. To facilitate the future investigation of the two populations of transcriptionally distinct Drd1-expressing MSNs, we next identified potential marker genes that are both differentially and highly expressed within a population. This analysis identified Ebf1 as a potential marker gene for the Drd1-MSN-1 population (Fig. 2e) and Htr4 as a potential marker gene for the Drd1-MSN-2 population (Fig. 2f). An additional characteristic of marker genes is that they are expressed in a large proportion of cells within the cluster. Plotting the expression of these genes on a UMAP shows differential enrichment of Ebf1 and Htr4 between the Drd1-expressing MSN populations (Fig. 2g). However, Ebf1 is also expressed in endothelial cells, while Htr4 is also expressed in Drd2and Drd3-MSNs. Thus, these genes may act as marker genes in combination with Drd1 but cannot act as marker genes alone.',\n",
       "   'entities': [{'entity': 'cocaine',\n",
       "     'start': 80,\n",
       "     'end': 87,\n",
       "     'label': 'SUBSTANCE'},\n",
       "    {'entity': 'Drd1-expressing MSNs',\n",
       "     'start': 150,\n",
       "     'end': 171,\n",
       "     'label': 'CELL_TYPE'},\n",
       "    {'entity': 'Drd1-MSN-1', 'start': 174, 'end': 184, 'label': 'CELL_TYPE'},\n",
       "    {'entity': 'Drd1-MSN-2', 'start': 189, 'end': 199, 'label': 'CELL_TYPE'},\n",
       "    {'entity': 'Ebf1', 'start': 380, 'end': 384, 'label': 'GENE'},\n",
       "    {'entity': 'D1-MSNs', 'start': 420, 'end': 427, 'label': 'CELL_TYPE'},\n",
       "    {'entity': 'Calb1', 'start': 493, 'end': 498, 'label': 'GENE'},\n",
       "    {'entity': 'Calbindin 1', 'start': 528, 'end': 539, 'label': 'PROTEIN'},\n",
       "    {'entity': 'NAc', 'start': 570, 'end': 573, 'label': 'BRAIN_REGION'},\n",
       "    {'entity': 'NAc core', 'start': 600, 'end': 608, 'label': 'BRAIN_REGION'},\n",
       "    {'entity': 'Htr4', 'start': 700, 'end': 704, 'label': 'GENE'},\n",
       "    {'entity': 'serotonin receptor 4',\n",
       "     'start': 730,\n",
       "     'end': 751,\n",
       "     'label': 'PROTEIN'},\n",
       "    {'entity': 'NAc shell', 'start': 840, 'end': 849, 'label': 'BRAIN_REGION'},\n",
       "    {'entity': 'Trpc3', 'start': 1200, 'end': 1205, 'label': 'GENE'},\n",
       "    {'entity': 'Cacng3', 'start': 1265, 'end': 1271, 'label': 'GENE'},\n",
       "    {'entity': 'Kcnc1', 'start': 1400, 'end': 1405, 'label': 'GENE'},\n",
       "    {'entity': 'Kcnc2', 'start': 1407, 'end': 1412, 'label': 'GENE'},\n",
       "    {'entity': 'Kcng3', 'start': 1414, 'end': 1419, 'label': 'GENE'},\n",
       "    {'entity': 'Drd2', 'start': 2200, 'end': 2204, 'label': 'GENE'},\n",
       "    {'entity': 'Drd3', 'start': 2205, 'end': 2209, 'label': 'GENE'},\n",
       "    {'entity': 'Drd1', 'start': 2226, 'end': 2230, 'label': 'GENE'}]},\n",
       "  {'text': 'We previously demonstrated that a single acute cocaine exposure differentially alters the transcriptome of molecularly defined cell types within the NAc (Savell et al., 2020). To begin understanding how repeated cocaine exposure alters the transcriptome of distinct NAc population, we injected male and female Sprague-Dawley rat with 20 mg/kg cocaine for 7 consecutive days (Fig. 1a, Supp. Fig. 1). Before directly comparing acute and repeated exposures, we first investigated cell type-specific transcriptional response to cocaine exposure, regardless of type or duration. To do this, we collapsed both datasets and used a pseudobulk approach to identify cocaine-dependent, cell type-specific differentially expressed gene (DEG; Supp. Table 2). This approach increased the number of replicates used for statistical comparison and circumvented previously identified issues with deprecated differential expression method previously used for snRNA-seq (Squair et al., 2021). Strikingly, the Drd1-MSN-1 population exhibited the most robust transcriptional response to cocaine, with 236 total DEGs following cocaine experience (184 upregulated, 52 downregulated; Fig. 4a,b). These upregulated DEGs include immediate early gene (IEG) such as Nr4a3, Tiparp, and Fosl2 (Fig. 4b). Recent evidence has demonstrated that astrocyte within the NAc also play a key role in the neuronal and behavioral response to dopamine and drug of abuse (Wang et al., 2021; Corkrum et al., 2020; Shelkar et al., 2022). Disruption of glutamatergic transport, release, and signaling is implicated in substance use disorder (Reissner and Kalivas, 2010; Kalivas, 2009), and astrocyte are heavily involved in the regulation of this process across the brain (Mahmoud et al., 2019). Interestingly, Gria1, an ionotropic AMPA receptor, is significantly decreased in astrocyte following cocaine exposure (Fig. 4c). Unlike the Drd1-MSN-1 population, astrocyte exhibited more gene with cocaine-dependent decrease in expression (Fig. 4c,g). While most DEGs in the astrocyte population are downregulated with cocaine, Pde10a, the gene encoding the cAMP-regulated phosphodiesterase 10A, is upregulated by cocaine in both the astrocyte and Sst-interneuron cluster (Fig. 4c,d). Previous research has demonstrated that Drd1-expressing MSN exhibit robust transcriptional change in response to cocaine (Savell et al., 2020; Bertran-Gonzalez et al., 2008; Zhang et al., 2006; Zhang et al., 2005). More specifically, snRNA-seq from the rat NAc following acute cocaine exposure demonstrated that D1-MSN have the most robust transcriptional response to cocaine when compared to other transcriptionally distinct NAc cell type (Savell et al., 2020). In this dataset, only one of the Drd1-expressing MSN population (Drd1-MSN-1) had a robust transcriptional response to cocaine. For example, Nr4a3, Fosb, Fosl2, and Tiparp are all significantly increased following cocaine exposure only in the Drd1-MSN-1 population (Fig. 4e). Next, we set out to investigate the relationship of cocaine-dependent DEG across cocaine exposure paradigm. To do this, we first conducted DEG analysis comparing cocaine and saline sample with exposure type ignored (i.e., all cocaine sample vs all saline sample), acute exposure only, and repeated exposure sample only. The log2 (Fold Change) value for all DEG was then concatenated and subject to hierarchical clustering. Overall, this analysis identified that many of the gene altered by acute cocaine are also altered by repeated cocaine in the same direction (Fig. 4f-h). However, for the Drd1-MSN-1 population some gene altered by acute cocaine are not altered by repeated cocaine. Furthermore, while the overall pattern shows concordant gene expression pattern, some cocaine-regulated gene within the astrocyte cluster demonstrate opposite change with an acute or repeated exposure. To estimate the degree of concordance between gene expression program altered by acute and repeated cocaine exposure in a threshold-free manner, we next used the rank-rank hypergeometric overlap (RRHO) algorithm (Cahill et al., 2018). To complete this analysis, we ranked gene list by the log2 (Fold Change), which was calculated by dividing the log-normalized expression value of cocaine cell by the log-normalized expression value of saline cell. This analysis identified that, on a cell type-specific basis, acute and repeated cocaine gene expression signature are primarily concordant. For example, the acute and repeated cocaine change for the Drd1-MSN-1, astrocyte, and Sst-interneuron population are significantly concordant for both upregulated and downregulated gene (Fig. 4i-l). However, the strength of this overlap and concordance is substantially different for these population.',\n",
       "   'entities': [{'entity': 'NAc',\n",
       "     'start': 120,\n",
       "     'end': 123,\n",
       "     'label': 'BRAIN_REGION'},\n",
       "    {'entity': 'Sprague-Dawley rat',\n",
       "     'start': 250,\n",
       "     'end': 268,\n",
       "     'label': 'ORGANISM'},\n",
       "    {'entity': 'Drd1-MSN-1', 'start': 800, 'end': 810, 'label': 'NEURON_TYPE'},\n",
       "    {'entity': 'Nr4a3', 'start': 1000, 'end': 1005, 'label': 'GENE'},\n",
       "    {'entity': 'Tiparp', 'start': 1007, 'end': 1013, 'label': 'GENE'},\n",
       "    {'entity': 'Fosl2', 'start': 1015, 'end': 1020, 'label': 'GENE'},\n",
       "    {'entity': 'astrocyte', 'start': 1200, 'end': 1209, 'label': 'CELL_TYPE'},\n",
       "    {'entity': 'Gria1', 'start': 1500, 'end': 1505, 'label': 'GENE'},\n",
       "    {'entity': 'Pde10a', 'start': 1800, 'end': 1806, 'label': 'GENE'},\n",
       "    {'entity': 'Sst-interneuron',\n",
       "     'start': 1900,\n",
       "     'end': 1914,\n",
       "     'label': 'NEURON_TYPE'},\n",
       "    {'entity': 'D1-MSN', 'start': 2200, 'end': 2206, 'label': 'NEURON_TYPE'},\n",
       "    {'entity': 'Fosb', 'start': 2500, 'end': 2504, 'label': 'GENE'},\n",
       "    {'entity': 'Fosl2', 'start': 2506, 'end': 2511, 'label': 'GENE'},\n",
       "    {'entity': 'Tiparp', 'start': 2513, 'end': 2519, 'label': 'GENE'}]},\n",
       "  {'text': 'Within the past few decades, a major focus of drug addiction research has been on the transcriptional and epigenetic changes occurring in the NAc following drug exposure. While these studies have been integral in our understanding of how drug experience shapes the brain, they have Fig. 5. Identification of conserved neuronal cell types across higher order mammals. a-b, Heatmap colored by the Pearsons correlation coefficient between rat, nonhuman primate, and human cell types. Correlation coefficients were calculated using t-statistic for rat MSN DEGs and their orthologs. c, Heatmaps colored by tstatistic for specific MSN marker genes in rat, non-human primate, and human. d, Estimated conservation of NAc neuronal cell types across rat, non-human primate, and human. primarily used techniques that ignore cell-type heterogeneity. Furthermore, although the NAc is fundamental in reward processing and the rat is a primary model organism used for reward-related research, few studies have comprehensively profiled the rat NAc at the transcriptional level. snRNA-seq provides the unique opportunity to study both the cell type-specific transcriptional responses to drugs of abuse, as well as the cell-type heterogeneity that exists within the NAc. In the present study, we report the integration of a previously published NAc snRNA-seq dataset collected following acute cocaine exposure (Savell et al., 2020) with a new dataset collected following repeated cocaine exposure. The resulting dataset contains 39,254 nuclei across 16 transcriptionally distinct populations (11 neuronal and 5 non-neuronal), which more than doubles the size of our previously published work. The addition of both biological and technical replicates, along with the addition of over 24,000 nuclei, allowed us to generate a comprehensive transcriptional atlas of the rat NAc and investigate cell type-specific transcriptional responses to cocaine. We were also interested in understanding the distribution of expression of genes encoding different receptor systems, which provides additional information regarding which cell types may respond to diverse classes of drugs. For example, there is significant renewed interest in the action of psychedelic drugs, many of which act on serotonin receptors. Htr2a, the gene encoding serotonin receptor type 2A and the expected target of many psychedelics, is expressed in a small proportion of Drd1-and Drd2-expressing MSNs, as well as Sst-interneurons (Supp. Fig. 5a). Thus, these populations of cells will most likely respond to classic psychedelics. There is also considerable interest in the distribution of expression of opioid receptors in the NAc as opioid-associated overdose deaths continue to increase (Hedegaard et al., 2021;Ciccarone, 2021). Previous research has demonstrated that 25  of cells containing the delta opioid receptor (Oprd1) in the NAc also contained Met 5 -Enkaphalin (Svingos et al., 1998), a peptide encoded by the gene Penk. Penk is a marker gene for Drd2-expressing MSNs, suggesting that the delta opioid receptor is expressed in this population of neurons. Our data confirms this observation, as Oprd1 is highly expressed in the Drd2-MSN-1 population (Supp. Fig. 5a-c). A major focus of this study was to characterize cell type-specific gene programs induced by cocaine experience. Prior work has revealed distinct differences in transcriptomic changes between volitional cocaine paradigms, such has self-administration, and experimenteradministered injections. For example, after a prolonged withdrawal period, animals that underwent cocaine self-administration exhibit increased surface expression of calcium permeable AMPA receptors, while animals that underwent experimenter-administered injections do not (McCutcheon et al., 2011). Additionally, a recent study systematically tested transcriptomic changes occurring in the VTA across several cocaine paradigms and identified several IEGs that are upregulated by experimenter-administered cocaine injections (Campbell et al., 2021). Like this study, we also identified several IEGs (Fosl2, Tiparp, Fosb), that are upregulated by experimenter-administered cocaine injections. Notably, we did not observe significant locomotor sensitization to repeated doses of cocaine in the present study (Supp. Fig. 1b). While sensitization is commonly (but not always (Gulley et al., 2003;Mayfield et al., 1992;Cass et al., 1993)) observed with repeated cocaine injections, it is possible that distinct gene expression patterns may be observed in animals that exhibit sensitized responses to cocaine as compared to those that do not. This will be an interesting consideration for future studies that are able to obtain single-nuclei transcriptional profiles while preserving individual animal variability in the data structure. Ultimately, the results presented in the current study shed light on cocaine-induced, cell-type specific transcriptomic changes occurring in the NAc, expanding our understanding of how drug experience shapes the brain. Further research investigating paradigmspecific transcriptional changes may identify molecular correlates of more complex behaviors, such as cocaine seeking or craving. The largest transcriptional response to cocaine experience occurred in the Drd1-MSN-1 population (Fig. 4a,b). These changes included IEGs such as Fosl2. IEGs and other cocaine-regulated genes in the Drd1-MSN-1 population are induced by both acute and repeated cocaine, but to a lesser extent with repeated cocaine. This data is in agreement with previous research that demonstrated induction of FOSB protein by repeated cocaine was weaker as compared to acute cocaine (Zhang et al., 2005). The second population of Drd1-expressing MSNs within the NAc, Drd1-MSN-2, did not have a robust transcriptional response to cocaine experience. A multitude of previous research has demonstrated that Drd1-expressing cells within the NAc exhibit transcriptional and epigenetic responses to drugs of abuse (Savell et al., 2020;Bertran-Gonzalez et al., 2008;Zhang et al., 2006;Zhang et al., 2005;Zipperly et al., 2021). Thus, the identification of a population of Drd1-expressing MSNs in the NAc with a significantly attenuated transcriptional response to cocaine is intriguing because it highlights a previously underappreciated level of molecular heterogeneity within this population of cells. Notably, further analysis of population-specific DEGs also revealed an enrichment of calcium transport-related terms for Drd1-MSN-1 and potassium transport-related terms for Drd1-MSN-2. The differential enrichment for calcium-related terms and genes may explain the transcriptional response to cocaine in the Drd1-MSN-1 population, as many IEGs rely on calcium-dependent signaling cascades for induction. Our dataset also provided the opportunity to identify marker genes for these two populations. We identified Ebf1 as a marker gene for the Drd1-MSN-1 population and Htr4 as a marker gene for the Drd1-MSN-2 population. In situ experiments validated the presence of two distinct populations of Drd1-expressing MSNs based on the presence of Htr4 and Ebf1. These in situ experiments also revealed a spatial separation of these MSNs, as Drd1Ebf1 cells were primarily found in the NAc core and Drd1Htr4 cells were primarily found in the NAc shell. Transcriptional similarity based on spatial location may also explain why the Drd1-MSN-1 population is more closely related to the Drd2-MSN-1 population than the Drd1-MSN-2 population (Supp. Fig. 4a). In fact, analysis of the relationship between cell types within the NHP identifies that Drd1-and Drd2-expressing MSNs from the same striatal subregions are more closely related than Drd1-expressing MSNs found in distinct striatal compartments (Supp. Fig. 4b). Similarly, the MSN.D1_A and MSN.D2_A human populations are more closely related to each other than other Drd1-or Drd2-expressing MSNs, although no spatial information is available in that dataset (Supp. Fig. 4c). Together, these results suggest that at least some component of transcriptional diversity in the NAc reflects the spatial organization of this structure. Although sex differences in drug-induced behavioral adaptations have long been appreciated (Hu et al., 2004;Bobzean et al., 2014;Becker and Chartoff, 2019;Evans, 2007), more recent research has also demonstrated sex-specific drug-induced molecular adaptations (López et al., 2021). Our dataset includes nuclei from both male and female rats, and thus we examined sex-specific transcriptional responses to cocaine across cell populations within the NAc, again using the collapsed acute and repeated dataset to enable pseudobulk treatment comparisons between male and females. In general, the direction of cocaine-induced mRNA changes was largely concordant between males and females across most cell populations (Supp. Fig. 6a-d), however significance criteria were often only met for one sex (Supp. Table 3). Overall, the Drd1-MSN-1 population was the most transcriptionally responsive to cocaine for both sexes in terms of total number of genes altered. Consistent with the sex-combined identification of cocaine DEGs, immediate early gene Nr4a3, transcription elongation factor Ell2, and cytoskeletal regulator Elmo2 were all significantly elevated in the Drd1-MSN-1 population in both sexes (Supp. Fig. 6e). Interestingly, sexspecific cocaine DEG comparisons revealed that while some genes, like Nr4a3 and Ell2, display relatively consistent patterns of expression between males and females across populations (Supp. Fig. 6e-f), other genes like Hs6st3 exhibit more dynamic patterns of expression (Supp. Fig. 6g). In conclusion, we have generated a comprehensive transcriptional atlas of the rat NAc that also provides information regarding cocainedependent gene expression changes. The use of snRNA-seq provided the ability to understand cell types that potentially mediate cellular and behavioral responses to drugs of abuse. Analysis of this dataset resulted in the identification of transcriptionally distinct populations of Drd1expressing MSNs that differ by their spatial location and transcriptional response to cocaine. This dataset will be a resource for future research into transcriptionally distinct populations of NAc cells and their responses to drugs of abuse.',\n",
       "   'entities': [{'entity': 'drug addiction',\n",
       "     'start': 38,\n",
       "     'end': 52,\n",
       "     'label': 'DISORDER'},\n",
       "    {'entity': 'NAc', 'start': 123, 'end': 126, 'label': 'BRAIN_REGION'},\n",
       "    {'entity': 'MSN', 'start': 400, 'end': 403, 'label': 'CELL_TYPE'},\n",
       "    {'entity': 'NAc', 'start': 500, 'end': 503, 'label': 'BRAIN_REGION'},\n",
       "    {'entity': 'snRNA-seq', 'start': 700, 'end': 709, 'label': 'TECHNIQUE'},\n",
       "    {'entity': 'NAc', 'start': 787, 'end': 790, 'label': 'BRAIN_REGION'},\n",
       "    {'entity': 'cocaine', 'start': 900, 'end': 907, 'label': 'DRUG'},\n",
       "    {'entity': 'Htr2a', 'start': 1400, 'end': 1405, 'label': 'GENE'},\n",
       "    {'entity': 'serotonin receptor type 2A',\n",
       "     'start': 1420,\n",
       "     'end': 1445,\n",
       "     'label': 'PROTEIN'},\n",
       "    {'entity': 'Drd1', 'start': 1490, 'end': 1494, 'label': 'GENE'},\n",
       "    {'entity': 'Drd2', 'start': 1500, 'end': 1504, 'label': 'GENE'},\n",
       "    {'entity': 'Sst-interneurons',\n",
       "     'start': 1510,\n",
       "     'end': 1526,\n",
       "     'label': 'CELL_TYPE'},\n",
       "    {'entity': 'Oprd1', 'start': 1700, 'end': 1705, 'label': 'GENE'},\n",
       "    {'entity': 'delta opioid receptor',\n",
       "     'start': 1710,\n",
       "     'end': 1730,\n",
       "     'label': 'PROTEIN'},\n",
       "    {'entity': 'Penk', 'start': 1780, 'end': 1784, 'label': 'GENE'},\n",
       "    {'entity': 'Drd2-MSN-1', 'start': 1850, 'end': 1860, 'label': 'CELL_TYPE'},\n",
       "    {'entity': 'Fosl2', 'start': 2200, 'end': 2205, 'label': 'GENE'},\n",
       "    {'entity': 'Tiparp', 'start': 2207, 'end': 2213, 'label': 'GENE'},\n",
       "    {'entity': 'Fosb', 'start': 2215, 'end': 2219, 'label': 'GENE'},\n",
       "    {'entity': 'VTA', 'start': 2300, 'end': 2303, 'label': 'BRAIN_REGION'},\n",
       "    {'entity': 'Drd1-MSN-1', 'start': 2800, 'end': 2810, 'label': 'CELL_TYPE'},\n",
       "    {'entity': 'Drd1-MSN-2', 'start': 2900, 'end': 2910, 'label': 'CELL_TYPE'},\n",
       "    {'entity': 'Ebf1', 'start': 3300, 'end': 3304, 'label': 'GENE'},\n",
       "    {'entity': 'Htr4', 'start': 3350, 'end': 3354, 'label': 'GENE'},\n",
       "    {'entity': 'NAc core',\n",
       "     'start': 3450,\n",
       "     'end': 3458,\n",
       "     'label': 'BRAIN_REGION'},\n",
       "    {'entity': 'NAc shell',\n",
       "     'start': 3480,\n",
       "     'end': 3489,\n",
       "     'label': 'BRAIN_REGION'},\n",
       "    {'entity': 'Drd2-MSN-1', 'start': 3550, 'end': 3560, 'label': 'CELL_TYPE'},\n",
       "    {'entity': 'Nr4a3', 'start': 4200, 'end': 4205, 'label': 'GENE'},\n",
       "    {'entity': 'Ell2', 'start': 4230, 'end': 4234, 'label': 'GENE'},\n",
       "    {'entity': 'Elmo2', 'start': 4260, 'end': 4265, 'label': 'GENE'},\n",
       "    {'entity': 'Hs6st3', 'start': 4350, 'end': 4356, 'label': 'GENE'}]},\n",
       "  {'text': 'All experiments were performed in accordance with the University of Alabama at Birmingham Institutional Animal Care and Use Committee. Male or female adult rats were intraperitoneally injected with Cocaine hydrocholoride at a dose of 20 mgkg or saline. Locomotor testing was performed as previously described (Savell et al., 2020). Animal undergoing acute cocaine exposure received a single dose, while those undergoing repeated cocaine exposure received one injection per day for 7 days.',\n",
       "   'entities': [{'entity': 'University of Alabama at Birmingham Institutional Animal Care and Use Committee',\n",
       "     'start': 50,\n",
       "     'end': 130,\n",
       "     'label': 'ORGANIZATION'},\n",
       "    {'entity': 'Male or female adult rats',\n",
       "     'start': 132,\n",
       "     'end': 156,\n",
       "     'label': 'SPECIES'},\n",
       "    {'entity': 'intraperitoneally',\n",
       "     'start': 162,\n",
       "     'end': 179,\n",
       "     'label': 'ROUTE_OF_ADMINISTRATION'},\n",
       "    {'entity': 'Cocaine hydrocholoride',\n",
       "     'start': 198,\n",
       "     'end': 219,\n",
       "     'label': 'DRUG'},\n",
       "    {'entity': '20 mgkg', 'start': 230, 'end': 237, 'label': 'DOSAGE'},\n",
       "    {'entity': 'saline', 'start': 242, 'end': 248, 'label': 'DRUG'},\n",
       "    {'entity': 'Locomotor testing', 'start': 250, 'end': 266, 'label': 'TEST'},\n",
       "    {'entity': 'Savell et al., 2020',\n",
       "     'start': 293,\n",
       "     'end': 310,\n",
       "     'label': 'REFERENCE'},\n",
       "    {'entity': 'acute cocaine exposure',\n",
       "     'start': 333,\n",
       "     'end': 354,\n",
       "     'label': 'TREATMENT_CONDITION'},\n",
       "    {'entity': 'single dose', 'start': 365, 'end': 376, 'label': 'DOSAGE'},\n",
       "    {'entity': 'repeated cocaine exposure',\n",
       "     'start': 400,\n",
       "     'end': 423,\n",
       "     'label': 'TREATMENT_CONDITION'},\n",
       "    {'entity': '7 days', 'start': 456, 'end': 462, 'label': 'DURATION'}]},\n",
       "  {'text': 'Nuclei dissociation for snRNA-seq was performed as previously described (Savell et al., 2020). Nuclei from 4 animals of the same sex, treatment group, and exposure length were pooled for tissue preparation and loaded into a single GEM (gel bead in emulsion) well.',\n",
       "   'entities': [{'entity': 'Nuclei dissociation',\n",
       "     'start': 0,\n",
       "     'end': 19,\n",
       "     'label': 'PROCEDURE'},\n",
       "    {'entity': 'snRNA-seq', 'start': 24, 'end': 33, 'label': 'TECHNIQUE'},\n",
       "    {'entity': 'Savell et al., 2020',\n",
       "     'start': 61,\n",
       "     'end': 78,\n",
       "     'label': 'REFERENCE'},\n",
       "    {'entity': 'Nuclei', 'start': 80, 'end': 86, 'label': 'BIOLOGICAL_ENTITY'},\n",
       "    {'entity': '4 animals', 'start': 92, 'end': 101, 'label': 'QUANTITY'},\n",
       "    {'entity': 'sex',\n",
       "     'start': 115,\n",
       "     'end': 118,\n",
       "     'label': 'BIOLOGICAL_ATTRIBUTE'},\n",
       "    {'entity': 'treatment group',\n",
       "     'start': 120,\n",
       "     'end': 135,\n",
       "     'label': 'EXPERIMENTAL_GROUP'},\n",
       "    {'entity': 'exposure length',\n",
       "     'start': 141,\n",
       "     'end': 156,\n",
       "     'label': 'EXPERIMENTAL_PARAMETER'},\n",
       "    {'entity': 'tissue preparation',\n",
       "     'start': 170,\n",
       "     'end': 187,\n",
       "     'label': 'PROCEDURE'},\n",
       "    {'entity': 'GEM (gel bead in emulsion)',\n",
       "     'start': 202,\n",
       "     'end': 228,\n",
       "     'label': 'TECHNICAL_TERM'}]},\n",
       "  {'text': \"Libraries were constructed according to manufacturer's instructions using the Chromium Single Cell 3 Library Construction Kit (10 Genomics, catalog no. 1000092), which uses version 3 chemistry for gene expression. Nuclei (15,655) from adult rat NAc after acute cocaine and nuclei (23,599) from adult rat NAc after repeated cocaine were sequenced on the Illumina NextSeq 500 at the UAB Heflin Genomics core to a depth of 39,000 reads per nuclei and 22,000 reads per nuclei, respectively. All raw fastq files were aligned to the Ensembl mRatBn7.2 (Rn7) genome using CellRanger (v6.1.2) with the associated Ensembl gene transfer format (gtf) file (version 105). Cell Ranger filtered output was analyzed with Seurat v4.0.4 (Hao et al., 2021) using R v4.0.2. Nuclei containing 200 features and 5% of reads mapping to the mitochondrial genome were removed from further analysis. Molecular count data from each GEM (gel bead in emulsion) well were then log-normalized with a scaling factor of 10,000. Following normalization, and before dimensionality reduction, all GEM wells were then integrated with FindIntegrationAnchors() and IntegrateData using 17 principal components (PCs) and a resolution value of 0.1 for both the acute and repeated dataset. Local Inverse Simpsons Index was calculated to ensure successful integration of all 8 GEM wells using lisi v1.0 with default parameters. Recent research has demonstrated that ambient RNA can be captured with single cell or nucleus with droplet-based single-cell single-nucleus technology. To avoid interpreting any signal associated with ambient RNA, we used SoupX v1.5.2 (Young and Behjati, 2020) to remove any signal associated with ambient RNA. Droplet-based single-cell single-nucleus technology is also susceptible to the capture of doublets, or two cells/nuclei being captured within a single droplet. Homotypic doublets, or doublets containing two cells of the same cell type, are difficult to identify and remove. However, heterotypic doublets, or doublets containing two cells of different cell types, can be identified as cells that have a combination of cell type-specific transcriptional signatures. Heterotypic doublets were not identified in our acute dataset. However, heterotypic doublets were identified and removed from the repeated dataset using DoubletFinder v2.0.3 (McGinnis et al., 2019). For differential expression testing, gene count was first summed across GEM wells within each cell type to create a pseudobulked gene expression matrix. Any gene with count <5 was removed from further analysis. Differential expression testing was then completed using pseudobulked gene expression matrix with DESeq2 v1.28.1 (Love et al., 2014) using the Likelihood Ratio Test (LRT). The dataset utilized in this paper is from two separate experiments and therefore represents two separate batches. Batch effect can substantially alter transcriptional profile and to control for this, dataset was included as a covariate in all differential expression testing. In addition to dataset, treatment was used as a covariate for differential expression testing across genetic sex and cell type. Including these variables as covariates allows us to ensure that all DEGs are due to treatment, cell type, or sex and not another confounding variable. Gini coefficient was calculated using the gini() function that is available in the reldist v1.6-6 package (Gini, 1921; Handcock and Morris, 1999). Rank-rank hypergeometric overlap (RRHO) was calculated using the log2 (Fold Change) of all genes regardless of adjusted p-value. RRHO heatmap was generated using RRHO2 v1.0 (Cahill et al., 2018; Plaisier et al., 2010). Molecular function gene ontology term was generated with DEGs expressed in 10% of the cluster using g:GOSt made available by gProfiler (Raudvere et al., 2019). For this tool, the statistical domain scope set to Only annotated gene and significance threshold set to Benjamini-Hochberg FDR. All R code for this analysis is available at https://github.com/Jeremy-Day-Lab/Phillips_2023_NAc.\",\n",
       "   'entities': [{'entity': 'Chromium Single Cell 3 Library Construction Kit',\n",
       "     'start': 70,\n",
       "     'end': 112,\n",
       "     'label': 'KIT'},\n",
       "    {'entity': '10 Genomics', 'start': 114, 'end': 125, 'label': 'COMPANY'},\n",
       "    {'entity': '1000092', 'start': 140, 'end': 147, 'label': 'CATALOG_NUMBER'},\n",
       "    {'entity': 'version 3 chemistry',\n",
       "     'start': 163,\n",
       "     'end': 180,\n",
       "     'label': 'CHEMISTRY'},\n",
       "    {'entity': 'Nuclei (15,655) from adult rat NAc after acute cocaine',\n",
       "     'start': 183,\n",
       "     'end': 234,\n",
       "     'label': 'SAMPLE'},\n",
       "    {'entity': 'nuclei (23,599) from adult rat NAc after repeated cocaine',\n",
       "     'start': 239,\n",
       "     'end': 293,\n",
       "     'label': 'SAMPLE'},\n",
       "    {'entity': 'Illumina NextSeq 500',\n",
       "     'start': 310,\n",
       "     'end': 330,\n",
       "     'label': 'INSTRUMENT'},\n",
       "    {'entity': 'UAB Heflin Genomics core',\n",
       "     'start': 335,\n",
       "     'end': 360,\n",
       "     'label': 'FACILITY'},\n",
       "    {'entity': '39,000 reads per nuclei',\n",
       "     'start': 375,\n",
       "     'end': 397,\n",
       "     'label': 'SEQUENCING_DEPTH'},\n",
       "    {'entity': '22,000 reads per nuclei',\n",
       "     'start': 402,\n",
       "     'end': 424,\n",
       "     'label': 'SEQUENCING_DEPTH'},\n",
       "    {'entity': 'Ensembl mRatBn7.2 (Rn7) genome',\n",
       "     'start': 450,\n",
       "     'end': 480,\n",
       "     'label': 'GENOME'},\n",
       "    {'entity': 'CellRanger (v6.1.2)',\n",
       "     'start': 485,\n",
       "     'end': 503,\n",
       "     'label': 'SOFTWARE'},\n",
       "    {'entity': 'Ensembl gene transfer format (gtf) file (version 105)',\n",
       "     'start': 520,\n",
       "     'end': 570,\n",
       "     'label': 'FILE'},\n",
       "    {'entity': 'Seurat v4.0.4', 'start': 600, 'end': 612, 'label': 'SOFTWARE'},\n",
       "    {'entity': 'R v4.0.2', 'start': 633, 'end': 641, 'label': 'SOFTWARE'},\n",
       "    {'entity': 'GEM (gel bead in emulsion)',\n",
       "     'start': 740,\n",
       "     'end': 766,\n",
       "     'label': 'TECHNOLOGY'},\n",
       "    {'entity': 'FindIntegrationAnchors()',\n",
       "     'start': 900,\n",
       "     'end': 923,\n",
       "     'label': 'FUNCTION'},\n",
       "    {'entity': 'IntegrateData', 'start': 925, 'end': 938, 'label': 'FUNCTION'},\n",
       "    {'entity': '17 principal components (PCs)',\n",
       "     'start': 943,\n",
       "     'end': 970,\n",
       "     'label': 'PARAMETER'},\n",
       "    {'entity': 'resolution value of 0.1',\n",
       "     'start': 975,\n",
       "     'end': 998,\n",
       "     'label': 'PARAMETER'},\n",
       "    {'entity': 'Local Inverse Simpsons Index',\n",
       "     'start': 1003,\n",
       "     'end': 1030,\n",
       "     'label': 'METRIC'},\n",
       "    {'entity': 'lisi v1.0', 'start': 1070, 'end': 1078, 'label': 'SOFTWARE'},\n",
       "    {'entity': 'SoupX v1.5.2',\n",
       "     'start': 1200,\n",
       "     'end': 1211,\n",
       "     'label': 'SOFTWARE'},\n",
       "    {'entity': 'DoubletFinder v2.0.3',\n",
       "     'start': 1500,\n",
       "     'end': 1518,\n",
       "     'label': 'SOFTWARE'},\n",
       "    {'entity': 'DESeq2 v1.28.1',\n",
       "     'start': 1700,\n",
       "     'end': 1713,\n",
       "     'label': 'SOFTWARE'},\n",
       "    {'entity': 'Likelihood Ratio Test (LRT)',\n",
       "     'start': 1720,\n",
       "     'end': 1745,\n",
       "     'label': 'TEST'},\n",
       "    {'entity': 'Gini coefficient',\n",
       "     'start': 2000,\n",
       "     'end': 2015,\n",
       "     'label': 'METRIC'},\n",
       "    {'entity': 'reldist v1.6-6',\n",
       "     'start': 2030,\n",
       "     'end': 2042,\n",
       "     'label': 'PACKAGE'},\n",
       "    {'entity': 'Rank-rank hypergeometric overlap (RRHO)',\n",
       "     'start': 2050,\n",
       "     'end': 2089,\n",
       "     'label': 'METHOD'},\n",
       "    {'entity': 'RRHO2 v1.0', 'start': 2150, 'end': 2160, 'label': 'SOFTWARE'},\n",
       "    {'entity': 'g:GOSt', 'start': 2300, 'end': 2306, 'label': 'TOOL'},\n",
       "    {'entity': 'gProfiler', 'start': 2330, 'end': 2339, 'label': 'SOFTWARE'},\n",
       "    {'entity': 'Benjamini-Hochberg FDR',\n",
       "     'start': 2400,\n",
       "     'end': 2421,\n",
       "     'label': 'METHOD'}]},\n",
       "  {'text': 'To correlate gene expression signatures across mammalian species, DEGs for rat cell types were first calculated using a pseudobulk approach. DEGs were then concatenated and t-statistic for every gene in every MSN subtype were calculated. The t-statistic was calculated with the following formula: Following calculation of t-statistics, the Pearson correlation coefficient was calculated using the cor() function in R. All R code for this analysis is available at https:github.comJeremy-Day-LabPhillips_2023_NAc.',\n",
       "   'entities': [{'entity': 'gene expression signatures',\n",
       "     'start': 15,\n",
       "     'end': 40,\n",
       "     'label': 'BIOMARKER'},\n",
       "    {'entity': 'mammalian species', 'start': 50, 'end': 66, 'label': 'TAXON'},\n",
       "    {'entity': 'DEGs', 'start': 72, 'end': 75, 'label': 'BIOMARKER'},\n",
       "    {'entity': 'rat cell types', 'start': 80, 'end': 93, 'label': 'CELL_TYPE'},\n",
       "    {'entity': 'pseudobulk approach',\n",
       "     'start': 110,\n",
       "     'end': 128,\n",
       "     'label': 'METHOD'},\n",
       "    {'entity': 'MSN subtype', 'start': 180, 'end': 191, 'label': 'CELL_TYPE'},\n",
       "    {'entity': 't-statistic', 'start': 197, 'end': 207, 'label': 'STATISTIC'},\n",
       "    {'entity': 'Pearson correlation coefficient',\n",
       "     'start': 250,\n",
       "     'end': 279,\n",
       "     'label': 'STATISTIC'},\n",
       "    {'entity': 'cor() function in R',\n",
       "     'start': 290,\n",
       "     'end': 308,\n",
       "     'label': 'METHOD'},\n",
       "    {'entity': 'https:github.comJeremy-Day-LabPhillips_2023_NAc',\n",
       "     'start': 337,\n",
       "     'end': 384,\n",
       "     'label': 'RESOURCE'}]},\n",
       "  {'text': 'Upon sacrifice, brains from 6 drug-naïve Sprague-Dawley rats (3F 3M) were extracted and immediately placed in isopentane at -40  C for 30 s. Tissue was stored at -80  C until day of sectioning. On day of sectioning, tissue was allowed to equilibrate at -20  C for 40 min and 10 μm coronal sections containing the NAc (1.92 mm to 1.44 mm from bregma) were collected using Leica cryostat and mounted on room temperature slides. Sections were air dried on dry ice for 60 min, then stored at -80  C. Sections were prepared for RNAscope Multiplex Fluorescent v2 Assay according to manufacturers protocol (Advanced Cell Diagnostics, 323110) (Wang et al., 2012). Probes included Htr4-C2 (544081-C2), and Drd1a-C3 (317031-C3) diluted 1:50 in Ebf1-C1 (891561-C1) and DAPI (320858). Associated Opal dyes were 690 (FP1497001KT) for Ebf1-C1, 570 (FP1488001KT) for Htr4-C2, and 520 (FP1487001KT) for Drd1a-C3 diluted 1:500 in TSA buffer. Sections were imaged using Keyence BZ-X800 microscope with OP-87767 filter cubes for GFP, Texas Red PE, DAPI, and Cy5 at 20 (PlanFluor NA 0.75)',\n",
       "   'entities': [{'entity': 'Sprague-Dawley rats',\n",
       "     'start': 33,\n",
       "     'end': 52,\n",
       "     'label': 'ORGANISM'},\n",
       "    {'entity': 'isopentane', 'start': 100, 'end': 110, 'label': 'CHEMICAL'},\n",
       "    {'entity': '-40  C', 'start': 114, 'end': 120, 'label': 'TEMPERATURE'},\n",
       "    {'entity': '-80  C', 'start': 160, 'end': 166, 'label': 'TEMPERATURE'},\n",
       "    {'entity': '-20  C', 'start': 240, 'end': 246, 'label': 'TEMPERATURE'},\n",
       "    {'entity': 'NAc', 'start': 300, 'end': 303, 'label': 'BRAIN_REGION'},\n",
       "    {'entity': '1.92 mm to 1.44 mm from bregma',\n",
       "     'start': 305,\n",
       "     'end': 335,\n",
       "     'label': 'COORDINATE'},\n",
       "    {'entity': 'Leica cryostat',\n",
       "     'start': 355,\n",
       "     'end': 368,\n",
       "     'label': 'EQUIPMENT'},\n",
       "    {'entity': 'RNAscope Multiplex Fluorescent v2 Assay',\n",
       "     'start': 450,\n",
       "     'end': 488,\n",
       "     'label': 'ASSAY'},\n",
       "    {'entity': 'Advanced Cell Diagnostics',\n",
       "     'start': 520,\n",
       "     'end': 544,\n",
       "     'label': 'ORGANIZATION'},\n",
       "    {'entity': '323110', 'start': 546, 'end': 552, 'label': 'PRODUCT_ID'},\n",
       "    {'entity': 'Wang et al., 2012',\n",
       "     'start': 555,\n",
       "     'end': 571,\n",
       "     'label': 'REFERENCE'},\n",
       "    {'entity': 'Htr4-C2', 'start': 592, 'end': 599, 'label': 'GENE_PROBE'},\n",
       "    {'entity': '544081-C2', 'start': 601, 'end': 610, 'label': 'PROBE_ID'},\n",
       "    {'entity': 'Drd1a-C3', 'start': 616, 'end': 624, 'label': 'GENE_PROBE'},\n",
       "    {'entity': '317031-C3', 'start': 626, 'end': 635, 'label': 'PROBE_ID'},\n",
       "    {'entity': 'Ebf1-C1', 'start': 653, 'end': 660, 'label': 'GENE_PROBE'},\n",
       "    {'entity': '891561-C1', 'start': 662, 'end': 671, 'label': 'PROBE_ID'},\n",
       "    {'entity': 'DAPI', 'start': 676, 'end': 680, 'label': 'STAIN'},\n",
       "    {'entity': '320858', 'start': 682, 'end': 688, 'label': 'PRODUCT_ID'},\n",
       "    {'entity': 'Opal dyes', 'start': 702, 'end': 711, 'label': 'FLUOROPHORE'},\n",
       "    {'entity': '690', 'start': 713, 'end': 716, 'label': 'DYE_WAVELENGTH'},\n",
       "    {'entity': 'FP1497001KT', 'start': 718, 'end': 729, 'label': 'PRODUCT_ID'},\n",
       "    {'entity': '570', 'start': 744, 'end': 747, 'label': 'DYE_WAVELENGTH'},\n",
       "    {'entity': 'FP1488001KT', 'start': 749, 'end': 760, 'label': 'PRODUCT_ID'},\n",
       "    {'entity': '520', 'start': 775, 'end': 778, 'label': 'DYE_WAVELENGTH'},\n",
       "    {'entity': 'FP1487001KT', 'start': 780, 'end': 791, 'label': 'PRODUCT_ID'},\n",
       "    {'entity': 'TSA buffer', 'start': 812, 'end': 822, 'label': 'BUFFER'},\n",
       "    {'entity': 'Keyence BZ-X800 microscope',\n",
       "     'start': 840,\n",
       "     'end': 865,\n",
       "     'label': 'EQUIPMENT'},\n",
       "    {'entity': 'OP-87767 filter cubes',\n",
       "     'start': 871,\n",
       "     'end': 891,\n",
       "     'label': 'EQUIPMENT'},\n",
       "    {'entity': 'GFP', 'start': 897, 'end': 900, 'label': 'FLUOROPHORE'},\n",
       "    {'entity': 'Texas Red PE',\n",
       "     'start': 902,\n",
       "     'end': 914,\n",
       "     'label': 'FLUOROPHORE'},\n",
       "    {'entity': 'DAPI', 'start': 916, 'end': 920, 'label': 'STAIN'},\n",
       "    {'entity': 'Cy5', 'start': 925, 'end': 928, 'label': 'FLUOROPHORE'},\n",
       "    {'entity': 'PlanFluor NA 0.75',\n",
       "     'start': 933,\n",
       "     'end': 950,\n",
       "     'label': 'LENS_SPEC'}]}]}"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "process_and_save_output_multiple(all_output, file_prefix = 'zero-shot-ner', prompt = ner_prompt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
